<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Visualize 'features' on a dimensional reduction plot — plot_feature • SCUBA</title><script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Visualize 'features' on a dimensional reduction plot — plot_feature"><meta name="description" content="Colors single cells on a dimensional reduction plot according to a 'feature'
(i.e. gene expression, PC scores, number of genes detected, etc.)"><meta property="og:description" content="Colors single cells on a dimensional reduction plot according to a 'feature'
(i.e. gene expression, PC scores, number of genes detected, etc.)"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">SCUBA</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.2.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/user_guide.html">User Guide</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch"><li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Visualize 'features' on a dimensional reduction plot</h1>

      <div class="d-none name"><code>plot_feature.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Colors single cells on a dimensional reduction plot according to a 'feature'
(i.e. gene expression, PC scores, number of genes detected, etc.)</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">plot_feature</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  <span class="va">features</span>,</span>
<span>  label_by <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  dims <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span>,</span>
<span>  cells <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  cols <span class="op">=</span> <span class="kw">if</span> <span class="op">(</span><span class="va">blend</span><span class="op">)</span> <span class="op">{</span></span>
<span>     <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"lightgrey"</span>, <span class="st">"#ff0000"</span>, <span class="st">"#00ff00"</span><span class="op">)</span></span>
<span> <span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>    </span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"lightgrey"</span>, <span class="st">"blue"</span><span class="op">)</span></span>
<span> <span class="op">}</span>,</span>
<span>  pt_size <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  order <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  min_cutoff <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  max_cutoff <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  reduction <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  split_by <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  keep_scale <span class="op">=</span> <span class="st">"feature"</span>,</span>
<span>  shape_by <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  slot <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  blend <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  blend_threshold <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  label <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  label_size <span class="op">=</span> <span class="fl">4</span>,</span>
<span>  label_color <span class="op">=</span> <span class="st">"black"</span>,</span>
<span>  repel <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  ncol <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  coord_fixed <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  by_col <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  combine <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  raster <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  raster_dpi <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">512</span>, <span class="fl">512</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-object">object<a class="anchor" aria-label="anchor" href="#arg-object"></a></dt>
<dd><p>a Seurat object or a SingleCellExperiment object</p></dd>


<dt id="arg-features">features<a class="anchor" aria-label="anchor" href="#arg-features"></a></dt>
<dd><p>Vector of features to plot. Features can come from:
- An <code>Assay</code> feature (e.g. a gene name - "MS4A1")
- A column name from meta.data (e.g. mitochondrial percentage - "percent.mito")
- A column name from a <code>DimReduc</code> object corresponding to the cell embedding values
(e.g. the PC 1 scores - "PC_1")</p></dd>


<dt id="arg-label-by">label_by<a class="anchor" aria-label="anchor" href="#arg-label-by"></a></dt>
<dd><p>A metadata column used for labeling groups on the featute plot,
if label is <code>TRUE</code>.</p></dd>


<dt id="arg-dims">dims<a class="anchor" aria-label="anchor" href="#arg-dims"></a></dt>
<dd><p>Dimensions to plot, must be a two-length numeric vector specifying x- and y-dimensions i.e. c(1,2) to plot the first and second dimensions from the
reduction results.</p></dd>


<dt id="arg-cells">cells<a class="anchor" aria-label="anchor" href="#arg-cells"></a></dt>
<dd><p>Vector of cells to plot (default is all cells)</p></dd>


<dt id="arg-cols">cols<a class="anchor" aria-label="anchor" href="#arg-cols"></a></dt>
<dd><p>The two colors to form the gradient over. Provide as string vector with
the first color corresponding to low values, the second to high. Also accepts a Brewer
color scale or vector of colors. Note: this will bin the data into number of colors provided.
When blend is <code>TRUE</code>, takes anywhere from 1-3 colors:</p><ul><li><p>1 color:</p><ul><li><p>Treated as color for double-negatives, will use default colors 2 and 3 for per-feature expression</p></li>
</ul></li>
<li><p>2 colors:</p><ul><li><p>Treated as colors for per-feature expression, will use default color 1 for double-negatives</p></li>
</ul></li>
<li><p>3+ colors:</p><ul><li><p>First color used for double-negatives, colors 2 and 3 used for per-feature expression, all others ignored</p></li>
</ul></li>
</ul></dd>


<dt id="arg-pt-size">pt_size<a class="anchor" aria-label="anchor" href="#arg-pt-size"></a></dt>
<dd><p>Adjust point size for plotting</p></dd>


<dt id="arg-order">order<a class="anchor" aria-label="anchor" href="#arg-order"></a></dt>
<dd><p>Boolean determining whether to plot cells in order of expression. Can be useful if
cells expressing given feature are getting buried.</p></dd>


<dt id="arg-min-cutoff-max-cutoff">min_cutoff, max_cutoff<a class="anchor" aria-label="anchor" href="#arg-min-cutoff-max-cutoff"></a></dt>
<dd><p>Vector of minimum and maximum cutoff values for each feature,
may specify quantile in the form of 'q##' where '##' is the quantile (eg, 'q1', 'q10')</p></dd>


<dt id="arg-reduction">reduction<a class="anchor" aria-label="anchor" href="#arg-reduction"></a></dt>
<dd><p>Which dimensionality reduction to use. If not specified, first searches for umap, then tsne, then pca</p></dd>


<dt id="arg-split-by">split_by<a class="anchor" aria-label="anchor" href="#arg-split-by"></a></dt>
<dd><p>A metadata column to split the feature plot by. Unlike <code><a href="https://satijalab.org/seurat/reference/FeaturePlot.html" class="external-link">Seurat::FeaturePlot</a></code>,
"ident" may not be passed since the ident functionality is not supported by SingleCellExperiment
objects. A metadata column name must be passed, or <code>NULL</code> to disable split plots.</p></dd>


<dt id="arg-keep-scale">keep_scale<a class="anchor" aria-label="anchor" href="#arg-keep-scale"></a></dt>
<dd><p>How to handle the color scale across multiple plots. Options are:</p><ul><li><p><code>feature</code> (default; by row/feature scaling):</p><ul><li><p>The plots for each individual feature are scaled to the maximum expression of the feature across the conditions provided to 'split_by'.</p></li>
</ul></li>
<li><p><code>all</code> (universal scaling):</p><ul><li><p>The plots for all features and conditions are scaled to the maximum expression value for the feature with the highest overall expression.</p></li>
</ul></li>
<li><p><code>NULL</code> (no scaling):</p><ul><li><p>Each individual plot is scaled to the maximum expression value of the feature in the condition provided to 'split_by'. Be aware setting NULL will result in color scales that are not comparable between plots.</p></li>
</ul></li>
</ul></dd>


<dt id="arg-shape-by">shape_by<a class="anchor" aria-label="anchor" href="#arg-shape-by"></a></dt>
<dd><p>If NULL, all points are circles (default). You can specify any
cell attribute (that can be pulled with FetchData) allowing for both
different colors and different shapes on cells.  Only applicable if <code>raster = FALSE</code>.</p></dd>


<dt id="arg-slot">slot<a class="anchor" aria-label="anchor" href="#arg-slot"></a></dt>
<dd><p>Which slot to pull expression data from? If <code>NULL</code>, defaults to "data" for Seurat objects, and "logcounts" for SingleCellExperiment objects.</p></dd>


<dt id="arg-blend">blend<a class="anchor" aria-label="anchor" href="#arg-blend"></a></dt>
<dd><p>Scale and blend expression values to visualize co-expression of two features</p></dd>


<dt id="arg-blend-threshold">blend_threshold<a class="anchor" aria-label="anchor" href="#arg-blend-threshold"></a></dt>
<dd><p>The color cutoff from weak signal to strong signal; ranges from 0 to 1.</p></dd>


<dt id="arg-label">label<a class="anchor" aria-label="anchor" href="#arg-label"></a></dt>
<dd><p>Whether to label the clusters</p></dd>


<dt id="arg-label-size">label_size<a class="anchor" aria-label="anchor" href="#arg-label-size"></a></dt>
<dd><p>Sets size of labels</p></dd>


<dt id="arg-label-color">label_color<a class="anchor" aria-label="anchor" href="#arg-label-color"></a></dt>
<dd><p>Sets the color of the label text</p></dd>


<dt id="arg-repel">repel<a class="anchor" aria-label="anchor" href="#arg-repel"></a></dt>
<dd><p>Repel labels</p></dd>


<dt id="arg-ncol">ncol<a class="anchor" aria-label="anchor" href="#arg-ncol"></a></dt>
<dd><p>Number of columns to combine multiple feature plots to, ignored if <code>split_by</code> is not <code>NULL</code></p></dd>


<dt id="arg-coord-fixed">coord_fixed<a class="anchor" aria-label="anchor" href="#arg-coord-fixed"></a></dt>
<dd><p>Plot cartesian coordinates with fixed aspect ratio</p></dd>


<dt id="arg-by-col">by_col<a class="anchor" aria-label="anchor" href="#arg-by-col"></a></dt>
<dd><p>If splitting by a factor, plot the splits per column with the features as rows; ignored if <code>blend = TRUE</code></p></dd>


<dt id="arg-combine">combine<a class="anchor" aria-label="anchor" href="#arg-combine"></a></dt>
<dd><p>Combine plots into a single <code><a href="https://patchwork.data-imaginist.com/reference/patchwork-package.html" class="external-link">patchwork</a>ed</code>
ggplot object. If <code>FALSE</code>, return a list of ggplot objects</p></dd>


<dt id="arg-raster">raster<a class="anchor" aria-label="anchor" href="#arg-raster"></a></dt>
<dd><p>Convert points to raster format, default is <code>NULL</code> which
automatically rasterizes if plotting more than 100,000 cells</p></dd>


<dt id="arg-raster-dpi">raster_dpi<a class="anchor" aria-label="anchor" href="#arg-raster-dpi"></a></dt>
<dd><p>Pixel resolution for rasterized plots, passed to geom_scattermore().
Default is c(512, 512).</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A <code><a href="https://patchwork.data-imaginist.com/reference/patchwork-package.html" class="external-link">patchwork</a>ed</code> ggplot object if
<code>combine = TRUE</code>; otherwise, a list of ggplot objects</p>
    </div>
    <div class="section level2">
    <h2 id="note">Note<a class="anchor" aria-label="anchor" href="#note"></a></h2>
    <p>For the old <code>do.hover</code> and <code>do.identify</code> functionality, please see
<code>HoverLocator</code> and <code>CellSelector</code>, respectively.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code>DimPlot</code> <code>HoverLocator</code>
<code>CellSelector</code></p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu">plot_feature</span><span class="op">(</span><span class="va">AML_Seurat</span>, features <span class="op">=</span> <span class="st">"PC_1"</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="plot_feature-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by William Showers, Jairav Desai, Clayton Smith, Craig Jordan, Austin Gillen.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

