<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>User Guide • SCUBA</title>
<script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="User Guide">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">SCUBA</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/user_guide.html">User Guide</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch">
<li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul>
</li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>User Guide</h1>
            
      

      <div class="d-none name"><code>user_guide.Rmd</code></div>
    </div>

    
    
<style type="text/css">
:root{--rs_blue: #217CA1}h4{margin-top:20px}.center{text-align:center;display:block}.white-text{color:white}.card-header{background-color:#CFB87C;margin-top:0px}.card-body{background-color:#E1E1E1;border-radius:0px 0px var(--bs-card-border-radius) var(--bs-card-border-radius)}.callout{padding:20px;margin:20px 0;border:1px solid #eee;border-left-width:8px;border-radius:8px}.callout h4,.callout h5{margin-top:0;margin-bottom:5px}.callout p:last-child{margin-bottom:0}.callout code{border-radius:3px}.callout+.bs-callout{margin-top:-5px}.callout-default{border-left-color:#777;background-color:#77735}.callout-default h4,.callout-default h5{color:#777}.callout-primary{border-left-color:#428bca;background-color:#428bca35}.callout-primary h4,.callout-primary h5{color:#428bca}.callout-success{border-left-color:#5cb85c;background-color:#5cb85c35}.callout-success h4,.callout-success h5{color:#5cb85c}.callout-danger{border-left-color:#d9534f;background-color:#d9534f35}.callout-danger h4,.callout-danger h5{color:#d9534f}.callout-warning{border-left-color:#f0ad4e;background-color:#f0ad4e35}.callout-warning h4,.callout-warning h5{color:#f0ad4e}.callout-info{border-left-color:#5bc0de;background-color:#5bc0de35}.callout-info h4,.callout-info h5{color:#5bc0de}.callout-bdc{border-left-color:#29527a;background-color:#29527a35}.callout-bdc h4,.callout-bdc h5{color:#29527a}.scroll-80{max-height:80px;overflow-y:auto;background-color:inherit}.scroll-100{max-height:100px;overflow-y:auto;background-color:inherit}.scroll-300{max-height:300px;overflow-y:auto;background-color:inherit}details{margin-bottom:10px}.header{background-color:#217CA1;color:#FFFFFF;border-radius:5px;padding:5px 15px;font-size:1.17em}.header span{margin-left:8px}.content{background-color:#E1E1E1;padding:15px}details[open] .header{border-radius:10px 10px 0px 0px}details[open] .content{border-radius:0px 0px 10px 10px}button.seurat{border:solid 1px;border-color:#000000;border-radius:10px 10px 0px 0px !important;background-color:#0072B299 !important;color:#000000 !important}button.seurat:hover{color:#FFFFFF !important}button.seurat.active{color:#FFFFFF !important;background-color:#0072B2 !important;border-bottom-color:#0072B2}div[id*=seurat] pre{background-color:#0072B211}button.singlecellexperiment{border:solid 1px;border-color:#000000;border-radius:10px 10px 0px 0px !important;background-color:#009E7399 !important;color:#000000 !important}button.singlecellexperiment:hover{color:#FFFFFF !important}button.singlecellexperiment.active{color:#FFFFFF !important;background-color:#009E73 !important;border-bottom-color:#009E73}div[id*=singlecellexperiment] pre{background-color:#009E7311}button.anndata{border:solid 1px;border-color:#000000;border-radius:10px 10px 0px 0px !important;background-color:#D55E0099 !important;color:#000000 !important}button.anndata:hover{color:#FFFFFF !important}button.anndata.active{color:#FFFFFF !important;background-color:#D55E00 !important;border-bottom-color:#D55E00}div[id*=anndata] pre{background-color:#D55E0011}button.mudata{border:solid 1px;border-color:#000000;border-radius:10px 10px 0px 0px !important;background-color:#FED10599 !important;color:#000000 !important}button.mudata:hover{color:#FFFFFF !important}button.mudata.active{color:#FFFFFF !important;background-color:#FED105 !important;border-bottom-color:#FED105}div[id*=mudata] pre{background-color:#FED10511}div.tab-content{border-color:#000000 !important;border-radius:0px 10px 10px 10px}.tab-content .tab-pane{display:none !important}.tab-content .active{display:block !important}
</style>
<script>
// Custom JavaScript for tabset/navbar panels
// Panels have classes for each object type, that need to map to both the <div>
// containing examples for each object, *and* the <button> tags in the tabset 
// pills that are used to change tabs.
// The default markdown syntax for tabset pills only applies classes to the 
// <div>. Here, it is also applied to the <button> tags.
$(document).ready(function(){
  // Find the tabset using the .tabset class and locate the decendent <button> 
  // tags. This only works on the pkgdown site (Rmd docs use <a> instead, and 
  // and do not appear to properly interperet JavaScript, at least in the R 
  // Studio browser)
  // .nav-pills.special-tabs: only apply classes to navbar menus with 
  // special-tabs class (there are other navbars on the pkgdown website)
  document.querySelectorAll(".nav-pills button").forEach(tab => {
      // console.log(tab)
       if (tab.textContent.trim() === "Seurat") {
           tab.classList.add("seurat");
       } else if (tab.textContent.trim() === "SingleCellExperiment"){
           tab.classList.add("singlecellexperiment");
       } else if (tab.textContent.trim() === "anndata") {
           tab.classList.add("anndata");
       }
   });
});

// document.addEventListener("DOMContentLoaded", function() {
//    
// });
</script><div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>SCUBA provides several functions for exploration, access, and
visualization of data from single-cell objects. Each of these is covered
here, for each object type supported by SCUBA. SCUBA currently supports
Seurat, SingleCellExperiment, and anndata objects.</p>
<div class="section level3">
<h3 id="naming-conventions">Naming Conventions<a class="anchor" aria-label="anchor" href="#naming-conventions"></a>
</h3>
<p>Single-cell object classes use different terms to refer to the the
same or similar concepts. For simplicity, this tutorial will use one
term for each concept. The concepts mentioned in the tutorial are given
below, and for each concept, the terms used for each object class are
mentioned, along with the term that will be used for this tutorial.</p>
<ul>
<li>
<strong>Sequencing modalities</strong>: the data structure used to
store each type of single-cell data (i.e. scRNA-seq, surface protein
measurements, ATAC-seq, etc.) is referred to as an “assay” in Seurat, an
“experiment” in SingleCellExperiment, and a “modality” in anndata. This
tutorial will use <em>“modality”</em>.</li>
<li>
<strong>Data structure for transformed data</strong>: in Seurat
objects, the structure used for storing raw counts, normalized counts,
and other transformations is known as a “slot”. In SingleCellExperiment,
this is referred to as an “assay”, and in anndata, a “layer”. This
tutorial will use <em>“layer”</em>.</li>
</ul>
<p>Please ensure you have loaded the SCUBA package before continuing.
See <a href="https://amc-heme.github.io/SCUBA/index.html#installation">here</a>
for instructions to install SCUBA.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load package</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://amc-heme.github.io/SCUBA/">SCUBA</a></span><span class="op">)</span></span></code></pre></div>
<!-- Starting with header 2 because header 1 tags in pkgdown keep the tabset pills 
     present in this document from rendering properly in pkgdown -->
<!-- https://github.com/r-lib/pkgdown/issues/2720 -->
<!--     ¯\_(ツ)_/¯     -->
</div>
</div>
<div class="section level2">
<h2 id="data-access">Data Access<a class="anchor" aria-label="anchor" href="#data-access"></a>
</h2>
<p>The primary data access method in SCUBA is <code><a href="../reference/fetch_data.html">fetch_data()</a></code>.
This function replicates the behavior of <code>FetchData()</code> in the
<code>SeuratObject</code> package. This function is an S3 generic that
executes methods based on the object type. When a Seurat object is
passed to <code><a href="../reference/fetch_data.html">fetch_data()</a></code>, the <code>FetchData</code> method
from the <code>SeuratObject</code> package is ran. When
SingleCellExperiment or anndata objects are passed, methods from SCUBA
are dispatched that replicate the behavior of <code>FetchData</code> in
these objects.</p>
<p><code><a href="../reference/fetch_data.html">fetch_data()</a></code> can be used to fetch either feature
expression data, metadata, or reduction coordinates, all of which are
fetched using the same parameter, <code>vars</code>. For each entry
passed to <code>vars</code>, <code><a href="../reference/fetch_data.html">fetch_data()</a></code> will
automatically identify the type of data matching the input and retrieve
the data accordingly.</p>
<div class="section level3">
<h3 id="fetching-expression-data">Fetching Expression Data<a class="anchor" aria-label="anchor" href="#fetching-expression-data"></a>
</h3>
<p><code><a href="../reference/fetch_data.html">fetch_data()</a></code> can retrieve feature expression data from
any assay in the object (or “experiment” in SingleCellExperiment and
“modality” in anndata objects). To retrieve feature expression data, you
can pass any number of features to the <code>vars</code> parameter.</p>
<p>Features may have the same name across different
assays/experiments/modalities. For example, there may be data for both
CD4 gene and surface protein expression, or there may be both gene
expression and chromatin accessibility data for the same gene. To ensure
feature data from the correct modality is returned,
<code><a href="../reference/fetch_data.html">fetch_data()</a></code> implements a syntax where a “key” is specified
before the feature name, with an underscore separating the key and the
feature name.</p>
<pre><code>{"key" of the modality} + "_" + {feature name}</code></pre>
<p>To determine the key to enter for the desired modality, we provide
the <code><a href="../reference/all_keys.html">all_keys()</a></code> utility function. The output is a named
character vector, with names equal to the name of the modality, and
values equal to the key to use.</p>
<!-- fetch_data expression data examples -->
<div class="section level4 tabset tabset-pills">
<h4 class="tabset tabset-pills" id="section"></h4>



<ul class="nav nav-pills" id="section" role="tablist">
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#seurat" id="seurat-tab" type="button" role="tab" aria-controls="seurat" aria-selected="true" class="active nav-link">Seurat</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#singlecellexperiment" id="singlecellexperiment-tab" type="button" role="tab" aria-controls="singlecellexperiment" aria-selected="false" class="nav-link">SingleCellExperiment</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#anndata" id="anndata-tab" type="button" role="tab" aria-controls="anndata" aria-selected="false" class="nav-link">anndata</button></li>
</ul>
<div class="tab-content">
<div class="active  tab-pane" id="seurat" role="tabpanel" aria-labelledby="seurat-tab">

<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/all_keys.html">all_keys</a></span><span class="op">(</span><span class="va">AML_Seurat</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## meta.data       RNA        AB       pca      umap </span></span>
<span><span class="co">##     "md_"    "rna_"     "ab_"     "PC_"   "UMAP_"</span></span></code></pre>
<p>In this example object, use the “rna” key to pull data from the “RNA”
assay.</p>
<p>To pull gene expression data for FLT3, for example, you would pass
“rna_FLT3” to <code>vars</code>.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/fetch_data.html">fetch_data</a></span><span class="op">(</span></span>
<span>  <span class="va">AML_Seurat</span>,</span>
<span>  vars <span class="op">=</span> <span class="st">"rna_FLT3"</span></span>
<span>  <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="co"># First 10 rows are shown for this example</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##          rna_FLT3</span></span>
<span><span class="co">## 487013_1 0.000000</span></span>
<span><span class="co">## 39207_1  0.000000</span></span>
<span><span class="co">## 861619_1 0.000000</span></span>
<span><span class="co">## 561110_1 0.000000</span></span>
<span><span class="co">## 283967_1 0.000000</span></span>
<span><span class="co">## 422573_1 0.000000</span></span>
<span><span class="co">## 453256_1 1.620299</span></span>
<span><span class="co">## 531766_1 0.000000</span></span>
<span><span class="co">## 796968_1 0.000000</span></span>
<span><span class="co">## 624345_1 2.037607</span></span></code></pre>
</div>
<div class="tab-pane" id="singlecellexperiment" role="tabpanel" aria-labelledby="singlecellexperiment-tab">

<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/all_keys.html">all_keys</a></span><span class="op">(</span><span class="fu"><a href="../reference/AML_SCE.html">AML_SCE</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##    RNA     AB    PCA   UMAP </span></span>
<span><span class="co">##  "RNA"   "AB"  "PCA" "UMAP"</span></span></code></pre>
<p>In this example object, use the “RNA” key to pull data from the “RNA”
assay.</p>
<p>To pull gene expression data for FLT3, for example, you would pass
“RNA_FLT3” to <code>vars</code>.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/fetch_data.html">fetch_data</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="../reference/AML_SCE.html">AML_SCE</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  vars <span class="op">=</span> <span class="st">"RNA_FLT3"</span></span>
<span>  <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="co"># First 10 rows are shown for this example</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##          RNA_FLT3</span></span>
<span><span class="co">## 487013_1 0.000000</span></span>
<span><span class="co">## 39207_1  0.000000</span></span>
<span><span class="co">## 861619_1 0.000000</span></span>
<span><span class="co">## 561110_1 0.000000</span></span>
<span><span class="co">## 283967_1 0.000000</span></span>
<span><span class="co">## 422573_1 0.000000</span></span>
<span><span class="co">## 453256_1 1.620299</span></span>
<span><span class="co">## 531766_1 0.000000</span></span>
<span><span class="co">## 796968_1 0.000000</span></span>
<span><span class="co">## 624345_1 2.037607</span></span></code></pre>
</div>
<div class="tab-pane" id="anndata" role="tabpanel" aria-labelledby="anndata-tab">

<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/all_keys.html">all_keys</a></span><span class="op">(</span><span class="fu"><a href="../reference/AML_h5ad.html">AML_h5ad</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##         X     X_pca    X_umap   protein </span></span>
<span><span class="co">##       "X"   "X_pca"  "X_umap" "protein"</span></span></code></pre>
<p>By convention, the gene expression modality in anndata objects is
named “X”. In this example object, use the “X” key to pull data from the
“RNA” assay.</p>
<p>To pull gene expression data for FLT3, for example, you would pass
“X_FLT3” to <code>vars</code>.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/fetch_data.html">fetch_data</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="../reference/AML_h5ad.html">AML_h5ad</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  vars <span class="op">=</span> <span class="st">"X_FLT3"</span></span>
<span>  <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="co"># First 10 rows are shown for this example</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##            X_FLT3</span></span>
<span><span class="co">## 487013_1 0.000000</span></span>
<span><span class="co">## 39207_1  0.000000</span></span>
<span><span class="co">## 861619_1 0.000000</span></span>
<span><span class="co">## 561110_1 0.000000</span></span>
<span><span class="co">## 283967_1 0.000000</span></span>
<span><span class="co">## 422573_1 0.000000</span></span>
<span><span class="co">## 453256_1 1.620299</span></span>
<span><span class="co">## 531766_1 0.000000</span></span>
<span><span class="co">## 796968_1 0.000000</span></span>
<span><span class="co">## 624345_1 2.037607</span></span></code></pre>
</div>
</div>
</div>
<div class="section level4 tabset tabset-pills">
<h4 class="tabset tabset-pills" id="section-1"></h4>
<p>Any number of features can be passed to vars as a character vector.
To view available features for a modality, SCUBA provides the
<code><a href="../reference/features_in_assay.html">features_in_assay()</a></code> generic. Use the <code>assay</code>
parameter to specify the modality, using the name instead of the key.
This utility generic can be used to query available features in a
modality, and check if a feature is represented in the data for the
assay before passage to fetch_data.</p>



<ul class="nav nav-pills" id="section-1" role="tablist">
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#seurat-1" id="seurat-1-tab" type="button" role="tab" aria-controls="seurat-1" aria-selected="true" class="active nav-link">Seurat</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#singlecellexperiment-1" id="singlecellexperiment-1-tab" type="button" role="tab" aria-controls="singlecellexperiment-1" aria-selected="false" class="nav-link">SingleCellExperiment</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#anndata-1" id="anndata-1-tab" type="button" role="tab" aria-controls="anndata-1" aria-selected="false" class="nav-link">anndata</button></li>
</ul>
<div class="tab-content">
<div class="active  tab-pane" id="seurat-1" role="tabpanel" aria-labelledby="seurat-1-tab">

<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">available_genes</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="../reference/features_in_assay.html">features_in_assay</a></span><span class="op">(</span></span>
<span>    <span class="va">AML_Seurat</span>,</span>
<span>    <span class="co"># For this object, "RNA" is the gene expression assay</span></span>
<span>    assay <span class="op">=</span> <span class="st">"RNA"</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="va">available_genes</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "ACTG1"   "ADGRG1"  "AHSP"    "AIF1"    "ANK1"    "ANKRD28" "ANLN"   </span></span>
<span><span class="co">##  [8] "ANP32E"  "ANPEP"   "ANXA2"</span></span></code></pre>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Test if a gene is present in the RNA assay before passage to fetch_data</span></span>
<span><span class="st">"MEIS1"</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">available_genes</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] TRUE</span></span></code></pre>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Request data for multiple features</span></span>
<span><span class="fu"><a href="../reference/fetch_data.html">fetch_data</a></span><span class="op">(</span></span>
<span>  <span class="va">AML_Seurat</span>,</span>
<span>  vars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"rna_FLT3"</span>, <span class="st">"rna_MEIS1"</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##          rna_FLT3 rna_MEIS1</span></span>
<span><span class="co">## 487013_1 0.000000         0</span></span>
<span><span class="co">## 39207_1  0.000000         0</span></span>
<span><span class="co">## 861619_1 0.000000         0</span></span>
<span><span class="co">## 561110_1 0.000000         0</span></span>
<span><span class="co">## 283967_1 0.000000         0</span></span>
<span><span class="co">## 422573_1 0.000000         0</span></span>
<span><span class="co">## 453256_1 1.620299         0</span></span>
<span><span class="co">## 531766_1 0.000000         0</span></span>
<span><span class="co">## 796968_1 0.000000         0</span></span>
<span><span class="co">## 624345_1 2.037607         0</span></span></code></pre>
</div>
<div class="tab-pane" id="singlecellexperiment-1" role="tabpanel" aria-labelledby="singlecellexperiment-1-tab">

<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">available_genes</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="../reference/features_in_assay.html">features_in_assay</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="../reference/AML_SCE.html">AML_SCE</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>    <span class="co"># In this object, RNA is the name of the gene expression experiment</span></span>
<span>    assay <span class="op">=</span> <span class="st">"RNA"</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="va">available_genes</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "ACTG1"   "ADGRG1"  "AHSP"    "AIF1"    "ANK1"    "ANKRD28" "ANLN"   </span></span>
<span><span class="co">##  [8] "ANP32E"  "ANPEP"   "ANXA2"</span></span></code></pre>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Test if a gene is present in the RNA assay before passage to fetch_data</span></span>
<span><span class="st">"MEIS1"</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">available_genes</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] TRUE</span></span></code></pre>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Request data for multiple features</span></span>
<span><span class="fu"><a href="../reference/fetch_data.html">fetch_data</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="../reference/AML_SCE.html">AML_SCE</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  vars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"RNA_FLT3"</span>, <span class="st">"RNA_MEIS1"</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##          RNA_FLT3 RNA_MEIS1</span></span>
<span><span class="co">## 487013_1 0.000000         0</span></span>
<span><span class="co">## 39207_1  0.000000         0</span></span>
<span><span class="co">## 861619_1 0.000000         0</span></span>
<span><span class="co">## 561110_1 0.000000         0</span></span>
<span><span class="co">## 283967_1 0.000000         0</span></span>
<span><span class="co">## 422573_1 0.000000         0</span></span>
<span><span class="co">## 453256_1 1.620299         0</span></span>
<span><span class="co">## 531766_1 0.000000         0</span></span>
<span><span class="co">## 796968_1 0.000000         0</span></span>
<span><span class="co">## 624345_1 2.037607         0</span></span></code></pre>
</div>
<div class="tab-pane" id="anndata-1" role="tabpanel" aria-labelledby="anndata-1-tab">

<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">available_genes</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="../reference/features_in_assay.html">features_in_assay</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="../reference/AML_h5ad.html">AML_h5ad</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>    assay <span class="op">=</span> <span class="st">"X"</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="va">available_genes</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "ACTG1"   "ADGRG1"  "AHSP"    "AIF1"    "ANK1"    "ANKRD28" "ANLN"   </span></span>
<span><span class="co">##  [8] "ANP32E"  "ANPEP"   "ANXA2"</span></span></code></pre>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Test if a gene is present in the RNA assay before passage to fetch_data</span></span>
<span><span class="st">"MEIS1"</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">available_genes</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] TRUE</span></span></code></pre>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Request data for multiple features</span></span>
<span><span class="fu"><a href="../reference/fetch_data.html">fetch_data</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="../reference/AML_h5ad.html">AML_h5ad</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  vars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"X_FLT3"</span>, <span class="st">"X_MEIS1"</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##            X_FLT3 X_MEIS1</span></span>
<span><span class="co">## 487013_1 0.000000       0</span></span>
<span><span class="co">## 39207_1  0.000000       0</span></span>
<span><span class="co">## 861619_1 0.000000       0</span></span>
<span><span class="co">## 561110_1 0.000000       0</span></span>
<span><span class="co">## 283967_1 0.000000       0</span></span>
<span><span class="co">## 422573_1 0.000000       0</span></span>
<span><span class="co">## 453256_1 1.620299       0</span></span>
<span><span class="co">## 531766_1 0.000000       0</span></span>
<span><span class="co">## 796968_1 0.000000       0</span></span>
<span><span class="co">## 624345_1 2.037607       0</span></span></code></pre>
<!-- {-} Ends tabset interface -->
</div>
</div>
</div>
<div class="section level4 unnumbered">
<h4 class="unnumbered" id="section-2"></h4>
</div>
</div>
<div class="section level3">
<h3 id="expression-data-from-alternate-layers">Expression Data from Alternate Layers<a class="anchor" aria-label="anchor" href="#expression-data-from-alternate-layers"></a>
</h3>
<p>By default, <code><a href="../reference/fetch_data.html">fetch_data()</a></code> will pull data from the
normalized counts layer. In Seurat objects, this is the “Data” layer and
in SingleCellExperiment objects, this is “logcounts”. In anndata
objects, it is customary to store normalized count data in the X matrix.
(If normalized counts data in your object is stored elsewhere, see the
“anndata” tab below).</p>
<p>To pull data from alternate layers, use the <code>layer</code>
parameter.</p>
<!-- Tabset interface uses an <h5> header since this is subordinate to the 
    <h4> section above -->
<div class="section level4 tabset tabset-pills">
<h4 class="tabset tabset-pills" id="section-3"></h4>
<!-- Tabset buttons are rendered properly as long as they are one header level
    subordinate of the .tabset element -->



<ul class="nav nav-pills" id="section-3" role="tablist">
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#seurat-2" id="seurat-2-tab" type="button" role="tab" aria-controls="seurat-2" aria-selected="true" class="active nav-link">Seurat</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#singlecellexperiment-2" id="singlecellexperiment-2-tab" type="button" role="tab" aria-controls="singlecellexperiment-2" aria-selected="false" class="nav-link">SingleCellExperiment</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#anndata-2" id="anndata-2-tab" type="button" role="tab" aria-controls="anndata-2" aria-selected="false" class="nav-link">anndata</button></li>
</ul>
<div class="tab-content">
<div class="active  tab-pane" id="seurat-2" role="tabpanel" aria-labelledby="seurat-2-tab">

<p>Below, feature expression data is pulled from raw counts (the
<code>counts</code> slot).</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/fetch_data.html">fetch_data</a></span><span class="op">(</span></span>
<span>  <span class="va">AML_Seurat</span>,</span>
<span>  vars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"rna_FLT3"</span>, <span class="st">"rna_MEIS1"</span><span class="op">)</span>,</span>
<span>  slot <span class="op">=</span> <span class="st">"counts"</span></span>
<span>  <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##          rna_FLT3 rna_MEIS1</span></span>
<span><span class="co">## 487013_1        0         0</span></span>
<span><span class="co">## 39207_1         0         0</span></span>
<span><span class="co">## 861619_1        0         0</span></span>
<span><span class="co">## 561110_1        0         0</span></span>
<span><span class="co">## 283967_1        0         0</span></span>
<span><span class="co">## 422573_1        0         0</span></span>
<span><span class="co">## 453256_1        3         0</span></span>
<span><span class="co">## 531766_1        0         0</span></span>
<span><span class="co">## 796968_1        0         0</span></span>
<span><span class="co">## 624345_1        4         0</span></span></code></pre>
<div class="callout callout-info">
<h4>
Note
</h4>
<p>
In Seurat v4 and earlier, the <code>layer</code> parameter was named
<code>slot</code>. If you are using Seurat v4 or earlier, you will need
to use <code>slot</code> instead of <code>layer</code>. This difference
does not affect SingleCellExperiment or anndata objects: for these
objects, only <code>layer</code> should be used.
</p>
</div>
</div>
<div class="tab-pane" id="singlecellexperiment-2" role="tabpanel" aria-labelledby="singlecellexperiment-2-tab">

<p>Below, feature expression data is pulled from raw counts (the
<code>counts</code> layer, referred to as the <code>counts</code>
“assay” in SingleCellExperiment).</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/fetch_data.html">fetch_data</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="../reference/AML_SCE.html">AML_SCE</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  vars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"RNA_FLT3"</span>, <span class="st">"RNA_MEIS1"</span><span class="op">)</span>,</span>
<span>  layer <span class="op">=</span> <span class="st">"counts"</span></span>
<span>  <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##          RNA_FLT3 RNA_MEIS1</span></span>
<span><span class="co">## 487013_1        0         0</span></span>
<span><span class="co">## 39207_1         0         0</span></span>
<span><span class="co">## 861619_1        0         0</span></span>
<span><span class="co">## 561110_1        0         0</span></span>
<span><span class="co">## 283967_1        0         0</span></span>
<span><span class="co">## 422573_1        0         0</span></span>
<span><span class="co">## 453256_1        3         0</span></span>
<span><span class="co">## 531766_1        0         0</span></span>
<span><span class="co">## 796968_1        0         0</span></span>
<span><span class="co">## 624345_1        4         0</span></span></code></pre>
</div>
<div class="tab-pane" id="anndata-2" role="tabpanel" aria-labelledby="anndata-2-tab">

<p>If the layer parameter is undefined for anndata objects,
<code><a href="../reference/fetch_data.html">fetch_data()</a></code> will use whichever matrix is present in X. If
normalized counts data is stored in an alternate layer in the
<code>layers</code> slot, the name of that layer must be passed to
<code>layers</code> to retrieve the normalized counts matrix.</p>
<p>In the example object below, normalized counts are stored in X, and
raw counts are stored in an alternate layer named “counts”.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Raw counts in "counts"</span></span>
<span><span class="fu"><a href="../reference/fetch_data.html">fetch_data</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="../reference/AML_h5ad.html">AML_h5ad</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  vars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"X_FLT3"</span>, <span class="st">"X_MEIS1"</span><span class="op">)</span>,</span>
<span>  layer <span class="op">=</span> <span class="st">"counts"</span></span>
<span>  <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##          X_FLT3 X_MEIS1</span></span>
<span><span class="co">## 487013_1      0       0</span></span>
<span><span class="co">## 39207_1       0       0</span></span>
<span><span class="co">## 861619_1      0       0</span></span>
<span><span class="co">## 561110_1      0       0</span></span>
<span><span class="co">## 283967_1      0       0</span></span>
<span><span class="co">## 422573_1      0       0</span></span>
<span><span class="co">## 453256_1      3       0</span></span>
<span><span class="co">## 531766_1      0       0</span></span>
<span><span class="co">## 796968_1      0       0</span></span>
<span><span class="co">## 624345_1      4       0</span></span></code></pre>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Normalized counts in X</span></span>
<span><span class="fu"><a href="../reference/fetch_data.html">fetch_data</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="../reference/AML_h5ad.html">AML_h5ad</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  vars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"X_FLT3"</span>, <span class="st">"X_MEIS1"</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##            X_FLT3 X_MEIS1</span></span>
<span><span class="co">## 487013_1 0.000000       0</span></span>
<span><span class="co">## 39207_1  0.000000       0</span></span>
<span><span class="co">## 861619_1 0.000000       0</span></span>
<span><span class="co">## 561110_1 0.000000       0</span></span>
<span><span class="co">## 283967_1 0.000000       0</span></span>
<span><span class="co">## 422573_1 0.000000       0</span></span>
<span><span class="co">## 453256_1 1.620299       0</span></span>
<span><span class="co">## 531766_1 0.000000       0</span></span>
<span><span class="co">## 796968_1 0.000000       0</span></span>
<span><span class="co">## 624345_1 2.037607       0</span></span></code></pre>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="fetching-metadata">Fetching Metadata<a class="anchor" aria-label="anchor" href="#fetching-metadata"></a>
</h3>
<p>To retrieve metadata, pass the name of the metadata variable to fetch
to the <code>vars</code> parameter of <code><a href="../reference/fetch_data.html">fetch_data()</a></code>. You can
pass any number of metadata variables to <code>vars</code> as a
character vector. Any metadata variable in the object can be retrieved,
and there are no differences in input between categorical and numeric
metadata variables.</p>
<p>To view available metadata variables in an object, SCUBA provides the
<code><a href="../reference/meta_varnames.html">meta_varnames()</a></code> utility function.</p>
<div class="section level4 tabset tabset-pills">
<h4 class="tabset tabset-pills" id="section-4"></h4>



<ul class="nav nav-pills" id="section-4" role="tablist">
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#seurat-3" id="seurat-3-tab" type="button" role="tab" aria-controls="seurat-3" aria-selected="true" class="active nav-link">Seurat</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#singlecellexperiment-3" id="singlecellexperiment-3-tab" type="button" role="tab" aria-controls="singlecellexperiment-3" aria-selected="false" class="nav-link">SingleCellExperiment</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#anndata-3" id="anndata-3-tab" type="button" role="tab" aria-controls="anndata-3" aria-selected="false" class="nav-link">anndata</button></li>
</ul>
<div class="tab-content">
<div class="active  tab-pane" id="seurat-3" role="tabpanel" aria-labelledby="seurat-3-tab">

<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># View available metadata</span></span>
<span><span class="fu"><a href="../reference/meta_varnames.html">meta_varnames</a></span><span class="op">(</span><span class="va">AML_Seurat</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "orig.ident"          "nCount_RNA"          "nFeature_RNA"       </span></span>
<span><span class="co">##  [4] "nCount_AB"           "nFeature_AB"         "nCount_BOTH"        </span></span>
<span><span class="co">##  [7] "nFeature_BOTH"       "BOTH_snn_res.0.9"    "seurat_clusters"    </span></span>
<span><span class="co">## [10] "Prediction_Ind"      "BOTH_snn_res.1"      "ClusterID"          </span></span>
<span><span class="co">## [13] "Batch"               "x"                   "y"                  </span></span>
<span><span class="co">## [16] "x_mean"              "y_mean"              "cor"                </span></span>
<span><span class="co">## [19] "ct"                  "prop"                "meandist"           </span></span>
<span><span class="co">## [22] "cDC"                 "B.cells"             "Myelocytes"         </span></span>
<span><span class="co">## [25] "Erythroid"           "Megakaryocte"        "Ident"              </span></span>
<span><span class="co">## [28] "RNA_snn_res.0.4"     "condensed_cell_type"</span></span></code></pre>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/fetch_data.html">fetch_data</a></span><span class="op">(</span></span>
<span>  <span class="va">AML_Seurat</span>,</span>
<span>  vars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"condensed_cell_type"</span>, <span class="st">"Batch"</span>, <span class="st">"nCount_RNA"</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="co"># head is used in this example since metadata will be returned for all cells</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                   condensed_cell_type    Batch nCount_RNA</span></span>
<span><span class="co">## 487013_1                 Plasma cells BM_200AB      10863</span></span>
<span><span class="co">## 39207_1                  Plasma cells BM_200AB       8403</span></span>
<span><span class="co">## 861619_1                 Plasma cells BM_200AB       8100</span></span>
<span><span class="co">## 561110_1                 Plasma cells BM_200AB       8151</span></span>
<span><span class="co">## 283967_1                    Primitive BM_200AB       8828</span></span>
<span><span class="co">## 422573_1                 Plasma cells BM_200AB       5188</span></span>
<span><span class="co">## 453256_1              Dendritic cells BM_200AB       7399</span></span>
<span><span class="co">## 531766_1                    Primitive BM_200AB       6621</span></span>
<span><span class="co">## 796968_1                    Primitive BM_200AB       6243</span></span>
<span><span class="co">## 624345_1 Plasmacytoid dendritic cells BM_200AB       5995</span></span></code></pre>
</div>
<div class="tab-pane" id="singlecellexperiment-3" role="tabpanel" aria-labelledby="singlecellexperiment-3-tab">

<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># View available metadata</span></span>
<span><span class="fu"><a href="../reference/meta_varnames.html">meta_varnames</a></span><span class="op">(</span><span class="fu"><a href="../reference/AML_SCE.html">AML_SCE</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "orig.ident"          "nCount_RNA"          "nFeature_RNA"       </span></span>
<span><span class="co">##  [4] "nCount_AB"           "nFeature_AB"         "nCount_BOTH"        </span></span>
<span><span class="co">##  [7] "nFeature_BOTH"       "BOTH_snn_res.0.9"    "seurat_clusters"    </span></span>
<span><span class="co">## [10] "Prediction_Ind"      "BOTH_snn_res.1"      "ClusterID"          </span></span>
<span><span class="co">## [13] "Batch"               "x"                   "y"                  </span></span>
<span><span class="co">## [16] "x_mean"              "y_mean"              "cor"                </span></span>
<span><span class="co">## [19] "ct"                  "prop"                "meandist"           </span></span>
<span><span class="co">## [22] "cDC"                 "B.cells"             "Myelocytes"         </span></span>
<span><span class="co">## [25] "Erythroid"           "Megakaryocte"        "Ident"              </span></span>
<span><span class="co">## [28] "RNA_snn_res.0.4"     "condensed_cell_type" "ident"</span></span></code></pre>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/fetch_data.html">fetch_data</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="../reference/AML_SCE.html">AML_SCE</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  vars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"condensed_cell_type"</span>, <span class="st">"Batch"</span>, <span class="st">"nCount_RNA"</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="co"># head is used in this example since metadata will be returned for all cells</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                   condensed_cell_type    Batch nCount_RNA</span></span>
<span><span class="co">## 487013_1                 Plasma cells BM_200AB      10863</span></span>
<span><span class="co">## 39207_1                  Plasma cells BM_200AB       8403</span></span>
<span><span class="co">## 861619_1                 Plasma cells BM_200AB       8100</span></span>
<span><span class="co">## 561110_1                 Plasma cells BM_200AB       8151</span></span>
<span><span class="co">## 283967_1                    Primitive BM_200AB       8828</span></span>
<span><span class="co">## 422573_1                 Plasma cells BM_200AB       5188</span></span>
<span><span class="co">## 453256_1              Dendritic cells BM_200AB       7399</span></span>
<span><span class="co">## 531766_1                    Primitive BM_200AB       6621</span></span>
<span><span class="co">## 796968_1                    Primitive BM_200AB       6243</span></span>
<span><span class="co">## 624345_1 Plasmacytoid dendritic cells BM_200AB       5995</span></span></code></pre>
</div>
<div class="tab-pane" id="anndata-3" role="tabpanel" aria-labelledby="anndata-3-tab">

<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># View available metadata</span></span>
<span><span class="fu"><a href="../reference/meta_varnames.html">meta_varnames</a></span><span class="op">(</span><span class="fu"><a href="../reference/AML_h5ad.html">AML_h5ad</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "nCount_RNA"          "nFeature_RNA"        "nCount_AB"          </span></span>
<span><span class="co">##  [4] "nFeature_AB"         "nCount_BOTH"         "nFeature_BOTH"      </span></span>
<span><span class="co">##  [7] "BOTH_snn_res.0.9"    "seurat_clusters"     "Prediction_Ind"     </span></span>
<span><span class="co">## [10] "BOTH_snn_res.1"      "ClusterID"           "Batch"              </span></span>
<span><span class="co">## [13] "x"                   "y"                   "x_mean"             </span></span>
<span><span class="co">## [16] "y_mean"              "cor"                 "ct"                 </span></span>
<span><span class="co">## [19] "prop"                "meandist"            "cDC"                </span></span>
<span><span class="co">## [22] "B.cells"             "Myelocytes"          "Erythroid"          </span></span>
<span><span class="co">## [25] "Megakaryocte"        "Ident"               "RNA_snn_res.0.4"    </span></span>
<span><span class="co">## [28] "condensed_cell_type"</span></span></code></pre>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/fetch_data.html">fetch_data</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="../reference/AML_h5ad.html">AML_h5ad</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  vars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"condensed_cell_type"</span>, <span class="st">"Batch"</span>, <span class="st">"nCount_RNA"</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="co"># head is used in this example since metadata will be returned for all cells</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                   condensed_cell_type    Batch nCount_RNA</span></span>
<span><span class="co">## 487013_1                 Plasma cells BM_200AB      10863</span></span>
<span><span class="co">## 39207_1                  Plasma cells BM_200AB       8403</span></span>
<span><span class="co">## 861619_1                 Plasma cells BM_200AB       8100</span></span>
<span><span class="co">## 561110_1                 Plasma cells BM_200AB       8151</span></span>
<span><span class="co">## 283967_1                    Primitive BM_200AB       8828</span></span>
<span><span class="co">## 422573_1                 Plasma cells BM_200AB       5188</span></span>
<span><span class="co">## 453256_1              Dendritic cells BM_200AB       7399</span></span>
<span><span class="co">## 531766_1                    Primitive BM_200AB       6621</span></span>
<span><span class="co">## 796968_1                    Primitive BM_200AB       6243</span></span>
<span><span class="co">## 624345_1 Plasmacytoid dendritic cells BM_200AB       5995</span></span></code></pre>
<!-- Empty section header ends tabset and keeps paragraph beneath from appearing
    under "anndata" tab -->
</div>
</div>
</div>
<div class="section level4 unnumbered">
<h4 class="unnumbered" id="section-5"></h4>
</div>
<div class="section level4">
<h4 id="fetch_metadata-generic">
<code>fetch_metadata</code> Generic<a class="anchor" aria-label="anchor" href="#fetch_metadata-generic"></a>
</h4>
<p>If you are retrieving just metadata from very large single-cell
objects, we offer a <code><a href="../reference/fetch_metadata.html">fetch_metadata()</a></code> method that is faster
than using <code><a href="../reference/fetch_data.html">fetch_data()</a></code>. As with <code><a href="../reference/fetch_data.html">fetch_data()</a></code>,
desired metadata variables are passed to <code>vars</code>.</p>
<div class="callout callout-info">
<h4>
Note
</h4>
<p>
In anndata objects, <code><a href="../reference/fetch_metadata.html">fetch_metadata()</a></code> does not offer
performance advantages over <code><a href="../reference/fetch_data.html">fetch_data()</a></code>.
</p>
</div>
</div>
<div class="section level4 tabset tabset-pills">
<h4 class="tabset tabset-pills" id="section-6"></h4>



<ul class="nav nav-pills" id="section-6" role="tablist">
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#seurat-4" id="seurat-4-tab" type="button" role="tab" aria-controls="seurat-4" aria-selected="true" class="active nav-link">Seurat</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#singlecellexperiment-4" id="singlecellexperiment-4-tab" type="button" role="tab" aria-controls="singlecellexperiment-4" aria-selected="false" class="nav-link">SingleCellExperiment</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#anndata-4" id="anndata-4-tab" type="button" role="tab" aria-controls="anndata-4" aria-selected="false" class="nav-link">anndata</button></li>
</ul>
<div class="tab-content">
<div class="active  tab-pane" id="seurat-4" role="tabpanel" aria-labelledby="seurat-4-tab">

<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/fetch_metadata.html">fetch_metadata</a></span><span class="op">(</span></span>
<span>  <span class="va">AML_Seurat</span>,</span>
<span>  vars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"condensed_cell_type"</span>, <span class="st">"Batch"</span>, <span class="st">"nCount_RNA"</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="co"># head is used in this example since metadata will be returned for all cells</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                   condensed_cell_type    Batch nCount_RNA</span></span>
<span><span class="co">## 487013_1                 Plasma cells BM_200AB      10863</span></span>
<span><span class="co">## 39207_1                  Plasma cells BM_200AB       8403</span></span>
<span><span class="co">## 861619_1                 Plasma cells BM_200AB       8100</span></span>
<span><span class="co">## 561110_1                 Plasma cells BM_200AB       8151</span></span>
<span><span class="co">## 283967_1                    Primitive BM_200AB       8828</span></span>
<span><span class="co">## 422573_1                 Plasma cells BM_200AB       5188</span></span>
<span><span class="co">## 453256_1              Dendritic cells BM_200AB       7399</span></span>
<span><span class="co">## 531766_1                    Primitive BM_200AB       6621</span></span>
<span><span class="co">## 796968_1                    Primitive BM_200AB       6243</span></span>
<span><span class="co">## 624345_1 Plasmacytoid dendritic cells BM_200AB       5995</span></span></code></pre>
</div>
<div class="tab-pane" id="singlecellexperiment-4" role="tabpanel" aria-labelledby="singlecellexperiment-4-tab">

<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/fetch_metadata.html">fetch_metadata</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="../reference/AML_SCE.html">AML_SCE</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  vars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"condensed_cell_type"</span>, <span class="st">"Batch"</span>, <span class="st">"nCount_RNA"</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="co"># head is used in this example since metadata will be returned for all cells</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                   condensed_cell_type    Batch nCount_RNA</span></span>
<span><span class="co">## 487013_1                 Plasma cells BM_200AB      10863</span></span>
<span><span class="co">## 39207_1                  Plasma cells BM_200AB       8403</span></span>
<span><span class="co">## 861619_1                 Plasma cells BM_200AB       8100</span></span>
<span><span class="co">## 561110_1                 Plasma cells BM_200AB       8151</span></span>
<span><span class="co">## 283967_1                    Primitive BM_200AB       8828</span></span>
<span><span class="co">## 422573_1                 Plasma cells BM_200AB       5188</span></span>
<span><span class="co">## 453256_1              Dendritic cells BM_200AB       7399</span></span>
<span><span class="co">## 531766_1                    Primitive BM_200AB       6621</span></span>
<span><span class="co">## 796968_1                    Primitive BM_200AB       6243</span></span>
<span><span class="co">## 624345_1 Plasmacytoid dendritic cells BM_200AB       5995</span></span></code></pre>
</div>
<div class="tab-pane" id="anndata-4" role="tabpanel" aria-labelledby="anndata-4-tab">

<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/fetch_metadata.html">fetch_metadata</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="../reference/AML_h5ad.html">AML_h5ad</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  vars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"condensed_cell_type"</span>, <span class="st">"Batch"</span>, <span class="st">"nCount_RNA"</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="co"># head is used in this example since metadata will be returned for all cells</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                   condensed_cell_type    Batch nCount_RNA</span></span>
<span><span class="co">## 487013_1                 Plasma cells BM_200AB      10863</span></span>
<span><span class="co">## 39207_1                  Plasma cells BM_200AB       8403</span></span>
<span><span class="co">## 861619_1                 Plasma cells BM_200AB       8100</span></span>
<span><span class="co">## 561110_1                 Plasma cells BM_200AB       8151</span></span>
<span><span class="co">## 283967_1                    Primitive BM_200AB       8828</span></span>
<span><span class="co">## 422573_1                 Plasma cells BM_200AB       5188</span></span>
<span><span class="co">## 453256_1              Dendritic cells BM_200AB       7399</span></span>
<span><span class="co">## 531766_1                    Primitive BM_200AB       6621</span></span>
<span><span class="co">## 796968_1                    Primitive BM_200AB       6243</span></span>
<span><span class="co">## 624345_1 Plasmacytoid dendritic cells BM_200AB       5995</span></span></code></pre>
<!-- Empty section header ends tabset and provides a slight margin between 
    the tabset and the next section header -->
</div>
</div>
</div>
<div class="section level4 unnumbered">
<h4 class="unnumbered" id="section-7"></h4>
<p>In a programming context, it may be useful to pull the full table.
This can be done with <code><a href="../reference/fetch_metadata.html">fetch_metadata()</a></code> using the
<code>full_table</code> parameter.</p>
</div>
<div class="section level4 tabset tabset-pills">
<h4 class="tabset tabset-pills" id="section-8"></h4>



<ul class="nav nav-pills" id="section-8" role="tablist">
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#seurat-5" id="seurat-5-tab" type="button" role="tab" aria-controls="seurat-5" aria-selected="true" class="active nav-link">Seurat</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#singlecellexperiment-5" id="singlecellexperiment-5-tab" type="button" role="tab" aria-controls="singlecellexperiment-5" aria-selected="false" class="nav-link">SingleCellExperiment</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#anndata-5" id="anndata-5-tab" type="button" role="tab" aria-controls="anndata-5" aria-selected="false" class="nav-link">anndata</button></li>
</ul>
<div class="tab-content">
<div class="active  tab-pane" id="seurat-5" role="tabpanel" aria-labelledby="seurat-5-tab">

<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">table</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="../reference/fetch_metadata.html">fetch_metadata</a></span><span class="op">(</span></span>
<span>    <span class="va">AML_Seurat</span>,</span>
<span>    full_table <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>    <span class="op">)</span></span>
<span></span>
<span><span class="va">table</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##             orig.ident nCount_RNA nFeature_RNA nCount_AB nFeature_AB</span></span>
<span><span class="co">## 487013_1 SeuratProject      10863          228     25709         195</span></span>
<span><span class="co">## 39207_1  SeuratProject       8403          210     31367         195</span></span>
<span><span class="co">## 861619_1 SeuratProject       8100          196     28166         195</span></span>
<span><span class="co">## 561110_1 SeuratProject       8151          179     14440         194</span></span>
<span><span class="co">## 283967_1 SeuratProject       8828          242      8203         191</span></span>
<span><span class="co">## 422573_1 SeuratProject       5188          147      8634         192</span></span>
<span><span class="co">## 453256_1 SeuratProject       7399          264     20649         194</span></span>
<span><span class="co">## 531766_1 SeuratProject       6621          232     17549         192</span></span>
<span><span class="co">## 796968_1 SeuratProject       6243          229     15178         193</span></span>
<span><span class="co">## 624345_1 SeuratProject       5995          246     22268         194</span></span></code></pre>
</div>
<div class="tab-pane" id="singlecellexperiment-5" role="tabpanel" aria-labelledby="singlecellexperiment-5-tab">

<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">table</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="../reference/fetch_metadata.html">fetch_metadata</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="../reference/AML_SCE.html">AML_SCE</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>    full_table <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>    <span class="op">)</span></span>
<span></span>
<span><span class="va">table</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 10 rows and 5 columns</span></span>
<span><span class="co">##             orig.ident nCount_RNA nFeature_RNA nCount_AB nFeature_AB</span></span>
<span><span class="co">##            &lt;character&gt;  &lt;numeric&gt;    &lt;integer&gt; &lt;numeric&gt;   &lt;integer&gt;</span></span>
<span><span class="co">## 487013_1 SeuratProject      10863          228     25709         195</span></span>
<span><span class="co">## 39207_1  SeuratProject       8403          210     31367         195</span></span>
<span><span class="co">## 861619_1 SeuratProject       8100          196     28166         195</span></span>
<span><span class="co">## 561110_1 SeuratProject       8151          179     14440         194</span></span>
<span><span class="co">## 283967_1 SeuratProject       8828          242      8203         191</span></span>
<span><span class="co">## 422573_1 SeuratProject       5188          147      8634         192</span></span>
<span><span class="co">## 453256_1 SeuratProject       7399          264     20649         194</span></span>
<span><span class="co">## 531766_1 SeuratProject       6621          232     17549         192</span></span>
<span><span class="co">## 796968_1 SeuratProject       6243          229     15178         193</span></span>
<span><span class="co">## 624345_1 SeuratProject       5995          246     22268         194</span></span></code></pre>
</div>
<div class="tab-pane" id="anndata-5" role="tabpanel" aria-labelledby="anndata-5-tab">

<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">table</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="../reference/fetch_metadata.html">fetch_metadata</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="../reference/AML_h5ad.html">AML_h5ad</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>    full_table <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>    <span class="op">)</span></span>
<span></span>
<span><span class="va">table</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##          nCount_RNA nFeature_RNA nCount_AB nFeature_AB nCount_BOTH</span></span>
<span><span class="co">## 487013_1      10863          228     25709         195       36572</span></span>
<span><span class="co">## 39207_1        8403          210     31367         195       39770</span></span>
<span><span class="co">## 861619_1       8100          196     28166         195       36266</span></span>
<span><span class="co">## 561110_1       8151          179     14440         194       22591</span></span>
<span><span class="co">## 283967_1       8828          242      8203         191       17031</span></span>
<span><span class="co">## 422573_1       5188          147      8634         192       13822</span></span>
<span><span class="co">## 453256_1       7399          264     20649         194       28048</span></span>
<span><span class="co">## 531766_1       6621          232     17549         192       24170</span></span>
<span><span class="co">## 796968_1       6243          229     15178         193       21421</span></span>
<span><span class="co">## 624345_1       5995          246     22268         194       28263</span></span></code></pre>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="fetching-reduction-coordinates">Fetching Reduction Coordinates<a class="anchor" aria-label="anchor" href="#fetching-reduction-coordinates"></a>
</h3>
<p>In a similar manner to expression data, reduction coordinates to
fetch are passed to <code>vars</code> using a “key” and underscore
system. In this case, you would enter the key of the reduction to use,
and an integer specifying the dimension of the reduction to pull
from.</p>
<pre><code>{"key" of the reduction} + "_" + {dimension}</code></pre>
<p>As with assays/modalities, run <code><a href="../reference/all_keys.html">all_keys()</a></code> to see the key
corresponding to the reduction you would like to pull data from.
Examples for each object type are provided below.</p>
<div class="section level4 tabset tabset-pills">
<h4 class="tabset tabset-pills" id="section-9"></h4>



<ul class="nav nav-pills" id="section-9" role="tablist">
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#seurat-6" id="seurat-6-tab" type="button" role="tab" aria-controls="seurat-6" aria-selected="true" class="active nav-link">Seurat</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#singlecellexperiment-6" id="singlecellexperiment-6-tab" type="button" role="tab" aria-controls="singlecellexperiment-6" aria-selected="false" class="nav-link">SingleCellExperiment</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#anndata-6" id="anndata-6-tab" type="button" role="tab" aria-controls="anndata-6" aria-selected="false" class="nav-link">anndata</button></li>
</ul>
<div class="tab-content">
<div class="active  tab-pane" id="seurat-6" role="tabpanel" aria-labelledby="seurat-6-tab">

<p>From running <code><a href="../reference/all_keys.html">all_keys()</a></code>, the key of the UMAP reduction
in this object is “UMAP_”.</p>
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/all_keys.html">all_keys</a></span><span class="op">(</span><span class="va">AML_Seurat</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## meta.data       RNA        AB       pca      umap </span></span>
<span><span class="co">##     "md_"    "rna_"     "ab_"     "PC_"   "UMAP_"</span></span></code></pre>
<p>To fetch the first and second dims, pass “UMAP_1” and “UMAP_2” to
<code>vars</code>.</p>
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/fetch_data.html">fetch_data</a></span><span class="op">(</span></span>
<span>  <span class="va">AML_Seurat</span>,</span>
<span>  vars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"UMAP_1"</span>, <span class="st">"UMAP_2"</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##             UMAP_1    UMAP_2</span></span>
<span><span class="co">## 487013_1 -1.643589  9.901035</span></span>
<span><span class="co">## 39207_1  -1.504856 10.125558</span></span>
<span><span class="co">## 861619_1 -1.448029 10.210141</span></span>
<span><span class="co">## 561110_1 -1.375756 10.512531</span></span>
<span><span class="co">## 283967_1 -1.405495  3.389042</span></span>
<span><span class="co">## 422573_1 -1.391848 10.582989</span></span>
<span><span class="co">## 453256_1 -2.676966  5.075310</span></span>
<span><span class="co">## 531766_1 -3.032746  5.115050</span></span>
<span><span class="co">## 796968_1 -3.034830  5.314629</span></span>
<span><span class="co">## 624345_1 -2.692851  4.533458</span></span></code></pre>
</div>
<div class="tab-pane" id="singlecellexperiment-6" role="tabpanel" aria-labelledby="singlecellexperiment-6-tab">

<p>From running <code><a href="../reference/all_keys.html">all_keys()</a></code>, the key of the UMAP reduction
in this object is “UMAP_”.</p>
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/all_keys.html">all_keys</a></span><span class="op">(</span><span class="fu"><a href="../reference/AML_SCE.html">AML_SCE</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##    RNA     AB    PCA   UMAP </span></span>
<span><span class="co">##  "RNA"   "AB"  "PCA" "UMAP"</span></span></code></pre>
<p>To fetch the first and second dims, pass “UMAP_1” and “UMAP_2” to
<code>vars</code>.</p>
<div class="sourceCode" id="cb72"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/fetch_data.html">fetch_data</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="../reference/AML_SCE.html">AML_SCE</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  vars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"UMAP_1"</span>, <span class="st">"UMAP_2"</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##             UMAP_1    UMAP_2</span></span>
<span><span class="co">## 487013_1 -1.643589  9.901035</span></span>
<span><span class="co">## 39207_1  -1.504856 10.125558</span></span>
<span><span class="co">## 861619_1 -1.448029 10.210141</span></span>
<span><span class="co">## 561110_1 -1.375756 10.512531</span></span>
<span><span class="co">## 283967_1 -1.405495  3.389042</span></span>
<span><span class="co">## 422573_1 -1.391848 10.582989</span></span>
<span><span class="co">## 453256_1 -2.676966  5.075310</span></span>
<span><span class="co">## 531766_1 -3.032746  5.115050</span></span>
<span><span class="co">## 796968_1 -3.034830  5.314629</span></span>
<span><span class="co">## 624345_1 -2.692851  4.533458</span></span></code></pre>
</div>
<div class="tab-pane" id="anndata-6" role="tabpanel" aria-labelledby="anndata-6-tab">

<p>From running <code><a href="../reference/all_keys.html">all_keys()</a></code>, the key of the UMAP reduction
in this object is “X_umap_”.</p>
<div class="sourceCode" id="cb74"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/all_keys.html">all_keys</a></span><span class="op">(</span><span class="fu"><a href="../reference/AML_h5ad.html">AML_h5ad</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##         X     X_pca    X_umap   protein </span></span>
<span><span class="co">##       "X"   "X_pca"  "X_umap" "protein"</span></span></code></pre>
<p>To fetch the first and second dims, pass “X_umap_1” and “X_umap_2” to
<code>vars</code>.</p>
<div class="sourceCode" id="cb76"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/fetch_data.html">fetch_data</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="../reference/AML_h5ad.html">AML_h5ad</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  vars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"X_umap_1"</span>, <span class="st">"X_umap_2"</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##           X_umap_1  X_umap_2</span></span>
<span><span class="co">## 487013_1 -1.643589  9.901035</span></span>
<span><span class="co">## 39207_1  -1.504856 10.125558</span></span>
<span><span class="co">## 861619_1 -1.448029 10.210141</span></span>
<span><span class="co">## 561110_1 -1.375756 10.512531</span></span>
<span><span class="co">## 283967_1 -1.405495  3.389042</span></span>
<span><span class="co">## 422573_1 -1.391848 10.582989</span></span>
<span><span class="co">## 453256_1 -2.676966  5.075310</span></span>
<span><span class="co">## 531766_1 -3.032746  5.115050</span></span>
<span><span class="co">## 796968_1 -3.034830  5.314629</span></span>
<span><span class="co">## 624345_1 -2.692851  4.533458</span></span></code></pre>
</div>
</div>
</div>
<div class="section level4 unnumbered">
<h4 class="unnumbered" id="section-10"></h4>
<p><br></p>
</div>
<div class="section level4">
<h4 id="fetch_reduction-generic">
<code>fetch_reduction</code> Generic<a class="anchor" aria-label="anchor" href="#fetch_reduction-generic"></a>
</h4>
<p>As with metadata, we provide a separate generic,
<code><a href="../reference/fetch_reduction.html">fetch_reduction()</a></code>, that is faster than using
<code><a href="../reference/fetch_data.html">fetch_data()</a></code> when retrieving just reduction coordinates
from very large single-cell objects. Performance gains are observed in
Seurat and SingleCellExperiment objects, but not in anndata objects
(performance is the same as <code><a href="../reference/fetch_data.html">fetch_data()</a></code> for this object
class).</p>
<p>The syntax for pulling reduction coordinates varies slightly from
<code><a href="../reference/fetch_data.html">fetch_data()</a></code>. Instead of entering the reduction name and
the dimensions together an underscore, this information is supplied to
separate parameters.</p>
<p>The key of reduction to pull data from is passed to
<code>reduction</code>, and the dimensions from which to pull
coordinates are passed in numeric format to <code>dims</code>. Any
number of dimensions can be passed, but in most cases, two dimensions
are pulled. If <code>dims</code> is not specified, the first and second
dimensions of the specified reduction are returned.</p>
</div>
<div class="section level4 tabset tabset-pills">
<h4 class="tabset tabset-pills" id="section-11"></h4>



<ul class="nav nav-pills" id="section-11" role="tablist">
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#seurat-7" id="seurat-7-tab" type="button" role="tab" aria-controls="seurat-7" aria-selected="true" class="active nav-link">Seurat</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#singlecellexperiment-7" id="singlecellexperiment-7-tab" type="button" role="tab" aria-controls="singlecellexperiment-7" aria-selected="false" class="nav-link">SingleCellExperiment</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#anndata-7" id="anndata-7-tab" type="button" role="tab" aria-controls="anndata-7" aria-selected="false" class="nav-link">anndata</button></li>
</ul>
<div class="tab-content">
<div class="active  tab-pane" id="seurat-7" role="tabpanel" aria-labelledby="seurat-7-tab">

<div class="sourceCode" id="cb78"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/fetch_reduction.html">fetch_reduction</a></span><span class="op">(</span></span>
<span>  <span class="va">AML_Seurat</span>,</span>
<span>  reduction <span class="op">=</span> <span class="st">"umap"</span>,</span>
<span>  dims <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="co"># head is used in this example since metadata will be returned for all cells</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##             UMAP_1    UMAP_2</span></span>
<span><span class="co">## 487013_1 -1.643589  9.901035</span></span>
<span><span class="co">## 39207_1  -1.504856 10.125558</span></span>
<span><span class="co">## 861619_1 -1.448029 10.210141</span></span>
<span><span class="co">## 561110_1 -1.375756 10.512531</span></span>
<span><span class="co">## 283967_1 -1.405495  3.389042</span></span>
<span><span class="co">## 422573_1 -1.391848 10.582989</span></span>
<span><span class="co">## 453256_1 -2.676966  5.075310</span></span>
<span><span class="co">## 531766_1 -3.032746  5.115050</span></span>
<span><span class="co">## 796968_1 -3.034830  5.314629</span></span>
<span><span class="co">## 624345_1 -2.692851  4.533458</span></span></code></pre>
</div>
<div class="tab-pane" id="singlecellexperiment-7" role="tabpanel" aria-labelledby="singlecellexperiment-7-tab">

<div class="sourceCode" id="cb80"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/fetch_reduction.html">fetch_reduction</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="../reference/AML_SCE.html">AML_SCE</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  reduction <span class="op">=</span> <span class="st">"UMAP"</span>,</span>
<span>  dims <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="co"># head is used in this example since metadata will be returned for all cells</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##             UMAP_1    UMAP_2</span></span>
<span><span class="co">## 487013_1 -1.643589  9.901035</span></span>
<span><span class="co">## 39207_1  -1.504856 10.125558</span></span>
<span><span class="co">## 861619_1 -1.448029 10.210141</span></span>
<span><span class="co">## 561110_1 -1.375756 10.512531</span></span>
<span><span class="co">## 283967_1 -1.405495  3.389042</span></span>
<span><span class="co">## 422573_1 -1.391848 10.582989</span></span>
<span><span class="co">## 453256_1 -2.676966  5.075310</span></span>
<span><span class="co">## 531766_1 -3.032746  5.115050</span></span>
<span><span class="co">## 796968_1 -3.034830  5.314629</span></span>
<span><span class="co">## 624345_1 -2.692851  4.533458</span></span></code></pre>
</div>
<div class="tab-pane" id="anndata-7" role="tabpanel" aria-labelledby="anndata-7-tab">

<div class="sourceCode" id="cb82"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/fetch_reduction.html">fetch_reduction</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="../reference/AML_h5ad.html">AML_h5ad</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  reduction <span class="op">=</span> <span class="st">"X_umap"</span>,</span>
<span>  dims <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="co"># head is used in this example since metadata will be returned for all cells</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##           X_umap_1  X_umap_2</span></span>
<span><span class="co">## 487013_1 -1.643589  9.901035</span></span>
<span><span class="co">## 39207_1  -1.504856 10.125558</span></span>
<span><span class="co">## 861619_1 -1.448029 10.210141</span></span>
<span><span class="co">## 561110_1 -1.375756 10.512531</span></span>
<span><span class="co">## 283967_1 -1.405495  3.389042</span></span>
<span><span class="co">## 422573_1 -1.391848 10.582989</span></span>
<span><span class="co">## 453256_1 -2.676966  5.075310</span></span>
<span><span class="co">## 531766_1 -3.032746  5.115050</span></span>
<span><span class="co">## 796968_1 -3.034830  5.314629</span></span>
<span><span class="co">## 624345_1 -2.692851  4.533458</span></span></code></pre>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="fetch-data-from-a-subset-of-cells">Fetch Data from a Subset of Cells<a class="anchor" aria-label="anchor" href="#fetch-data-from-a-subset-of-cells"></a>
</h3>
<p>In some circumstances, you may wish to access data from a subset of
cells, rather than the full object. <code><a href="../reference/fetch_data.html">fetch_data()</a></code>,
<code><a href="../reference/fetch_metadata.html">fetch_metadata()</a></code>, and <code><a href="../reference/fetch_reduction.html">fetch_reduction()</a></code>
provide a <code>cells</code> parameter to do this.</p>
<p>The <code>cells</code> parameter accepts a character vector with the
cell IDs for which to return data. We provide a utility function,
<code><a href="../reference/fetch_cells.html">fetch_cells()</a></code>, to generate this.</p>
<p><code><a href="../reference/fetch_cells.html">fetch_cells()</a></code> accepts a categorical metadata variable
and a vector of values/levels to filter by. Cells with values matching
the values entered will be returned.</p>
<p>To view available values for a metadata variable, you can use the
<code><a href="../reference/unique_values.html">unique_values()</a></code> function.</p>
<div class="section level4 tabset tabset-pills">
<h4 class="tabset tabset-pills" id="section-12"></h4>



<ul class="nav nav-pills" id="section-12" role="tablist">
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#seurat-8" id="seurat-8-tab" type="button" role="tab" aria-controls="seurat-8" aria-selected="true" class="active nav-link">Seurat</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#singlecellexperiment-8" id="singlecellexperiment-8-tab" type="button" role="tab" aria-controls="singlecellexperiment-8" aria-selected="false" class="nav-link">SingleCellExperiment</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#anndata-8" id="anndata-8-tab" type="button" role="tab" aria-controls="anndata-8" aria-selected="false" class="nav-link">anndata</button></li>
</ul>
<div class="tab-content">
<div class="active  tab-pane" id="seurat-8" role="tabpanel" aria-labelledby="seurat-8-tab">

<p>Identifying available values for the variable “Batch”:</p>
<div class="sourceCode" id="cb84"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/unique_values.html">unique_values</a></span><span class="op">(</span><span class="va">AML_Seurat</span>, var <span class="op">=</span> <span class="st">"Batch"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "BM_200AB"   "PBMC_200AB"</span></span></code></pre>
<p>Returning cells are returned where “Batch” is equal to
“PBMC_200AB”:</p>
<div class="sourceCode" id="cb86"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cell_subset</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="../reference/fetch_cells.html">fetch_cells</a></span><span class="op">(</span></span>
<span>    object <span class="op">=</span> <span class="va">AML_Seurat</span>, </span>
<span>    meta_var <span class="op">=</span> <span class="st">"Batch"</span>, </span>
<span>    meta_levels <span class="op">=</span> <span class="st">"PBMC_200AB"</span></span>
<span>    <span class="op">)</span></span>
<span></span>
<span><span class="co"># Preview cell IDs returned</span></span>
<span><span class="va">cell_subset</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "844178_2" "809777_2" "613075_2" "401992_2" "15150_2"  "208437_2"</span></span></code></pre>
<div class="sourceCode" id="cb88"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Number of cells</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">cell_subset</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 98</span></span></code></pre>
<p>Retrieving expression data for the subset:</p>
<div class="sourceCode" id="cb90"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="../reference/fetch_data.html">fetch_data</a></span><span class="op">(</span></span>
<span>    <span class="va">AML_Seurat</span>,</span>
<span>    vars <span class="op">=</span> <span class="st">"rna_FLT3"</span>,</span>
<span>    cells <span class="op">=</span> <span class="va">cell_subset</span></span>
<span>    <span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 'data.frame':    98 obs. of  1 variable:</span></span>
<span><span class="co">##  $ rna_FLT3: num  0 0 0 0 0 0 0 0 0 0 ...</span></span></code></pre>
<div class="sourceCode" id="cb92"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Compare with full data</span></span>
<span><span class="va">full_data</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="../reference/fetch_data.html">fetch_data</a></span><span class="op">(</span></span>
<span>    <span class="va">AML_Seurat</span>,</span>
<span>    vars <span class="op">=</span> <span class="st">"rna_FLT3"</span></span>
<span>    <span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">full_data</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 'data.frame':    250 obs. of  1 variable:</span></span>
<span><span class="co">##  $ rna_FLT3: num  0 0 0 0 0 ...</span></span></code></pre>
</div>
<div class="tab-pane" id="singlecellexperiment-8" role="tabpanel" aria-labelledby="singlecellexperiment-8-tab">

<p>Identifying available values for the variable “Batch”:</p>
<div class="sourceCode" id="cb94"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/unique_values.html">unique_values</a></span><span class="op">(</span><span class="fu"><a href="../reference/AML_SCE.html">AML_SCE</a></span><span class="op">(</span><span class="op">)</span>, var <span class="op">=</span> <span class="st">"Batch"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "BM_200AB"   "PBMC_200AB"</span></span></code></pre>
<p>Returning cells are returned where “Batch” is equal to
“PBMC_200AB”:</p>
<div class="sourceCode" id="cb96"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cell_subset</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="../reference/fetch_cells.html">fetch_cells</a></span><span class="op">(</span></span>
<span>    object <span class="op">=</span> <span class="fu"><a href="../reference/AML_SCE.html">AML_SCE</a></span><span class="op">(</span><span class="op">)</span>, </span>
<span>    meta_var <span class="op">=</span> <span class="st">"Batch"</span>, </span>
<span>    meta_levels <span class="op">=</span> <span class="st">"PBMC_200AB"</span></span>
<span>    <span class="op">)</span></span>
<span></span>
<span><span class="co"># Preview cell IDs returned</span></span>
<span><span class="va">cell_subset</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "844178_2" "809777_2" "613075_2" "401992_2" "15150_2"  "208437_2"</span></span></code></pre>
<div class="sourceCode" id="cb98"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Number of cells</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">cell_subset</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 98</span></span></code></pre>
<p>Retrieving expression data for the subset:</p>
<div class="sourceCode" id="cb100"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="../reference/fetch_data.html">fetch_data</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="../reference/AML_SCE.html">AML_SCE</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>    vars <span class="op">=</span> <span class="st">"RNA_FLT3"</span>,</span>
<span>    cells <span class="op">=</span> <span class="va">cell_subset</span></span>
<span>    <span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 'data.frame':    98 obs. of  1 variable:</span></span>
<span><span class="co">##  $ RNA_FLT3: num  0 0 0 0 0 0 0 0 0 0 ...</span></span></code></pre>
<div class="sourceCode" id="cb102"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Compare with full data</span></span>
<span><span class="va">full_data</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="../reference/fetch_data.html">fetch_data</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="../reference/AML_SCE.html">AML_SCE</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>    vars <span class="op">=</span> <span class="st">"RNA_FLT3"</span></span>
<span>    <span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">full_data</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 'data.frame':    250 obs. of  1 variable:</span></span>
<span><span class="co">##  $ RNA_FLT3: num  0 0 0 0 0 ...</span></span></code></pre>
</div>
<div class="tab-pane" id="anndata-8" role="tabpanel" aria-labelledby="anndata-8-tab">

<p>Identifying available values for the variable “Batch”:</p>
<div class="sourceCode" id="cb104"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/unique_values.html">unique_values</a></span><span class="op">(</span><span class="fu"><a href="../reference/AML_h5ad.html">AML_h5ad</a></span><span class="op">(</span><span class="op">)</span>, var <span class="op">=</span> <span class="st">"Batch"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] BM_200AB   PBMC_200AB</span></span>
<span><span class="co">## Levels: BM_200AB PBMC_200AB</span></span></code></pre>
<p>Returning cells are returned where “Batch” is equal to
“PBMC_200AB”:</p>
<div class="sourceCode" id="cb106"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cell_subset</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="../reference/fetch_cells.html">fetch_cells</a></span><span class="op">(</span></span>
<span>    object <span class="op">=</span> <span class="fu"><a href="../reference/AML_h5ad.html">AML_h5ad</a></span><span class="op">(</span><span class="op">)</span>, </span>
<span>    meta_var <span class="op">=</span> <span class="st">"Batch"</span>, </span>
<span>    meta_levels <span class="op">=</span> <span class="st">"PBMC_200AB"</span></span>
<span>    <span class="op">)</span></span>
<span></span>
<span><span class="co"># Preview cell IDs returned</span></span>
<span><span class="va">cell_subset</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "844178_2" "809777_2" "613075_2" "401992_2" "15150_2"  "208437_2"</span></span></code></pre>
<div class="sourceCode" id="cb108"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Number of cells</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">cell_subset</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 98</span></span></code></pre>
<p>Retrieving expression data for the subset:</p>
<div class="sourceCode" id="cb110"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="../reference/fetch_data.html">fetch_data</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="../reference/AML_h5ad.html">AML_h5ad</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>    vars <span class="op">=</span> <span class="st">"X_FLT3"</span>,</span>
<span>    cells <span class="op">=</span> <span class="va">cell_subset</span></span>
<span>    <span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 'data.frame':    98 obs. of  1 variable:</span></span>
<span><span class="co">##  $ X_FLT3: num  0 0 0 0 0 0 0 0 0 0 ...</span></span>
<span><span class="co">##  - attr(*, "pandas.index")=Index(['844178_2', '809777_2', '613075_2', '401992_2', '15150_2', '208437_2',</span></span>
<span><span class="co">##        '368193_2', '457499_2', '29093_2', '744754_2', '800098_2', '754060_2',</span></span>
<span><span class="co">##        '632183_2', '759680_2', '694449_2', '700832_2', '1257_2', '448980_2',</span></span>
<span><span class="co">##        '199181_2', '833067_2', '251571_2', '736115_2', '712624_2', '707144_2',</span></span>
<span><span class="co">##        '859209_2', '253883_2', '653923_2', '762295_2', '320896_2', '99676_2',</span></span>
<span><span class="co">##        '800149_2', '682908_2', '780178_2', '852293_2', '615431_2', '218771_2',</span></span>
<span><span class="co">##        '116904_2', '290694_2', '611301_2', '147803_2', '576873_2', '884381_2',</span></span>
<span><span class="co">##        '127725_2', '502119_2', '607265_2', '680163_2', '787312_2', '60207_2',</span></span>
<span><span class="co">##        '326911_2', '632894_2', '683104_2', '765082_2', '628813_2', '668067_2',</span></span>
<span><span class="co">##        '242368_2', '157340_2', '351199_2', '409871_2', '511910_2', '488361_2',</span></span>
<span><span class="co">##        '467352_2', '376467_2', '749121_2', '50361_2', '155312_2', '785055_2',</span></span>
<span><span class="co">##        '246318_2', '665215_2', '634705_2', '610060_2', '89949_2', '874939_2',</span></span>
<span><span class="co">##        '624530_2', '721485_2', '48168_2', '257195_2', '391063_2', '655228_2',</span></span>
<span><span class="co">##        '752608_2', '687706_2', '382710_2', '700441_2', '38920_2', '564255_2',</span></span>
<span><span class="co">##        '125842_2', '63869_2', '677793_2', '355446_2', '883406_2', '662718_2',</span></span>
<span><span class="co">##        '691696_2', '431743_2', '158371_2', '679107_2', '844492_2', '729807_2',</span></span>
<span><span class="co">##        '545562_2', '849364_2'],</span></span>
<span><span class="co">##       dtype='object')</span></span></code></pre>
<div class="sourceCode" id="cb112"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Compare with full data</span></span>
<span><span class="va">full_data</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="../reference/fetch_data.html">fetch_data</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="../reference/AML_h5ad.html">AML_h5ad</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>    vars <span class="op">=</span> <span class="st">"X_FLT3"</span></span>
<span>    <span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">full_data</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 'data.frame':    250 obs. of  1 variable:</span></span>
<span><span class="co">##  $ X_FLT3: num  0 0 0 0 0 ...</span></span>
<span><span class="co">##  - attr(*, "pandas.index")=Index(['487013_1', '39207_1', '861619_1', '561110_1', '283967_1', '422573_1',</span></span>
<span><span class="co">##        '453256_1', '531766_1', '796968_1', '624345_1',</span></span>
<span><span class="co">##        ...</span></span>
<span><span class="co">##        '883406_2', '662718_2', '691696_2', '431743_2', '158371_2', '679107_2',</span></span>
<span><span class="co">##        '844492_2', '729807_2', '545562_2', '849364_2'],</span></span>
<span><span class="co">##       dtype='object', length=250)</span></span></code></pre>
<!-- Empty header to end tabset panel and keep content below from appearing in
    anndata section -->
</div>
</div>
</div>
<div class="section level4 unnumbered">
<h4 class="unnumbered" id="section-13"></h4>
<div class="callout callout-info">
<h4>
Note
</h4>
<p>
<code><a href="../reference/fetch_cells.html">fetch_cells()</a></code> currently only returns a subset based on
values/levels of a single categorical variable. The ability to create
more complex filters will be added in the future.
</p>
</div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="data-exploration">Data Exploration<a class="anchor" aria-label="anchor" href="#data-exploration"></a>
</h2>
<p>SCUBA provides several utility functions that facilitate exploration
of objects before analysis.</p>
<div class="section level3">
<h3 id="view-overall-metadata-present">View Overall Metadata Present<a class="anchor" aria-label="anchor" href="#view-overall-metadata-present"></a>
</h3>
<p>It is often useful to view the names of all metadata variables in the
object, either as a first pass when exploring an object, or to recall
the name of a metadata variable before passage to
<code><a href="../reference/fetch_data.html">fetch_data()</a></code> or <code><a href="../reference/fetch_metadata.html">fetch_metadata()</a></code>. This is done
via <code><a href="../reference/meta_varnames.html">meta_varnames()</a></code>.</p>
<div class="section level4 tabset tabset-pills special-tabs">
<h4 class="tabset tabset-pills special-tabs" id="section-14"></h4>



<ul class="nav nav-pills" id="section-14" role="tablist">
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#seurat-9" id="seurat-9-tab" type="button" role="tab" aria-controls="seurat-9" aria-selected="true" class="active nav-link">Seurat</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#singlecellexperiment-9" id="singlecellexperiment-9-tab" type="button" role="tab" aria-controls="singlecellexperiment-9" aria-selected="false" class="nav-link">SingleCellExperiment</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#anndata-9" id="anndata-9-tab" type="button" role="tab" aria-controls="anndata-9" aria-selected="false" class="nav-link">anndata</button></li>
</ul>
<div class="tab-content">
<div class="active  tab-pane" id="seurat-9" role="tabpanel" aria-labelledby="seurat-9-tab">

<div class="sourceCode" id="cb114"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/meta_varnames.html">meta_varnames</a></span><span class="op">(</span></span>
<span>  <span class="va">AML_Seurat</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "orig.ident"          "nCount_RNA"          "nFeature_RNA"       </span></span>
<span><span class="co">##  [4] "nCount_AB"           "nFeature_AB"         "nCount_BOTH"        </span></span>
<span><span class="co">##  [7] "nFeature_BOTH"       "BOTH_snn_res.0.9"    "seurat_clusters"    </span></span>
<span><span class="co">## [10] "Prediction_Ind"      "BOTH_snn_res.1"      "ClusterID"          </span></span>
<span><span class="co">## [13] "Batch"               "x"                   "y"                  </span></span>
<span><span class="co">## [16] "x_mean"              "y_mean"              "cor"                </span></span>
<span><span class="co">## [19] "ct"                  "prop"                "meandist"           </span></span>
<span><span class="co">## [22] "cDC"                 "B.cells"             "Myelocytes"         </span></span>
<span><span class="co">## [25] "Erythroid"           "Megakaryocte"        "Ident"              </span></span>
<span><span class="co">## [28] "RNA_snn_res.0.4"     "condensed_cell_type"</span></span></code></pre>
</div>
<div class="tab-pane" id="singlecellexperiment-9" role="tabpanel" aria-labelledby="singlecellexperiment-9-tab">

<div class="sourceCode" id="cb116"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/meta_varnames.html">meta_varnames</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="../reference/AML_SCE.html">AML_SCE</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "orig.ident"          "nCount_RNA"          "nFeature_RNA"       </span></span>
<span><span class="co">##  [4] "nCount_AB"           "nFeature_AB"         "nCount_BOTH"        </span></span>
<span><span class="co">##  [7] "nFeature_BOTH"       "BOTH_snn_res.0.9"    "seurat_clusters"    </span></span>
<span><span class="co">## [10] "Prediction_Ind"      "BOTH_snn_res.1"      "ClusterID"          </span></span>
<span><span class="co">## [13] "Batch"               "x"                   "y"                  </span></span>
<span><span class="co">## [16] "x_mean"              "y_mean"              "cor"                </span></span>
<span><span class="co">## [19] "ct"                  "prop"                "meandist"           </span></span>
<span><span class="co">## [22] "cDC"                 "B.cells"             "Myelocytes"         </span></span>
<span><span class="co">## [25] "Erythroid"           "Megakaryocte"        "Ident"              </span></span>
<span><span class="co">## [28] "RNA_snn_res.0.4"     "condensed_cell_type" "ident"</span></span></code></pre>
</div>
<div class="tab-pane" id="anndata-9" role="tabpanel" aria-labelledby="anndata-9-tab">

<div class="sourceCode" id="cb118"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/meta_varnames.html">meta_varnames</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="../reference/AML_h5ad.html">AML_h5ad</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "nCount_RNA"          "nFeature_RNA"        "nCount_AB"          </span></span>
<span><span class="co">##  [4] "nFeature_AB"         "nCount_BOTH"         "nFeature_BOTH"      </span></span>
<span><span class="co">##  [7] "BOTH_snn_res.0.9"    "seurat_clusters"     "Prediction_Ind"     </span></span>
<span><span class="co">## [10] "BOTH_snn_res.1"      "ClusterID"           "Batch"              </span></span>
<span><span class="co">## [13] "x"                   "y"                   "x_mean"             </span></span>
<span><span class="co">## [16] "y_mean"              "cor"                 "ct"                 </span></span>
<span><span class="co">## [19] "prop"                "meandist"            "cDC"                </span></span>
<span><span class="co">## [22] "B.cells"             "Myelocytes"          "Erythroid"          </span></span>
<span><span class="co">## [25] "Megakaryocte"        "Ident"               "RNA_snn_res.0.4"    </span></span>
<span><span class="co">## [28] "condensed_cell_type"</span></span></code></pre>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="summarize-a-metadata-variable">Summarize a Metadata Variable<a class="anchor" aria-label="anchor" href="#summarize-a-metadata-variable"></a>
</h3>
<p>To explore a metadata variable in greater detail, use
<code><a href="../reference/unique_values.html">unique_values()</a></code>. This will print all values within the
metadata variable. This is done below with the “condensed_cell_type”
variable identified from <code><a href="../reference/meta_varnames.html">meta_varnames()</a></code>.</p>
<p>To view the unique values for a variable, pass the name of the
variable to the <code>var</code> parameter of the
<code><a href="../reference/unique_values.html">unique_values()</a></code> function.</p>
<div class="section level4 tabset tabset-pills">
<h4 class="tabset tabset-pills" id="section-15"></h4>



<ul class="nav nav-pills" id="section-15" role="tablist">
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#seurat-10" id="seurat-10-tab" type="button" role="tab" aria-controls="seurat-10" aria-selected="true" class="active nav-link">Seurat</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#singlecellexperiment-10" id="singlecellexperiment-10-tab" type="button" role="tab" aria-controls="singlecellexperiment-10" aria-selected="false" class="nav-link">SingleCellExperiment</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#anndata-10" id="anndata-10-tab" type="button" role="tab" aria-controls="anndata-10" aria-selected="false" class="nav-link">anndata</button></li>
</ul>
<div class="tab-content">
<div class="active  tab-pane" id="seurat-10" role="tabpanel" aria-labelledby="seurat-10-tab">

<div class="sourceCode" id="cb120"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/unique_values.html">unique_values</a></span><span class="op">(</span></span>
<span>  <span class="va">AML_Seurat</span>,</span>
<span>  var <span class="op">=</span> <span class="st">"condensed_cell_type"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre class="seurat"><code><span><span class="co">##  [1] "Plasma cells"                 "Primitive"                   </span></span>
<span><span class="co">##  [3] "Dendritic cells"              "Plasmacytoid dendritic cells"</span></span>
<span><span class="co">##  [5] "BM Monocytes"                 "NK Cells"                    </span></span>
<span><span class="co">##  [7] "CD8+ T Cells"                 "B Cells"                     </span></span>
<span><span class="co">##  [9] "CD4+ T Cells"                 "PBMC Monocytes"</span></span></code></pre>
</div>
<div class="tab-pane" id="singlecellexperiment-10" role="tabpanel" aria-labelledby="singlecellexperiment-10-tab">

<div class="sourceCode" id="cb122"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/unique_values.html">unique_values</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="../reference/AML_SCE.html">AML_SCE</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  var <span class="op">=</span> <span class="st">"condensed_cell_type"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb123"><pre class="sourceCode sce"><code class="sourceCode sci"><span id="cb123-1"><a href="#cb123-1" tabindex="-1"></a>##  [<span class="dv">1</span>] <span class="st">"Plasma cells"</span>                 <span class="st">"Primitive"</span>                   </span>
<span id="cb123-2"><a href="#cb123-2" tabindex="-1"></a>##  [<span class="dv">3</span>] <span class="st">"Dendritic cells"</span>              <span class="st">"Plasmacytoid dendritic cells"</span></span>
<span id="cb123-3"><a href="#cb123-3" tabindex="-1"></a>##  [<span class="dv">5</span>] <span class="st">"BM Monocytes"</span>                 <span class="st">"NK Cells"</span>                    </span>
<span id="cb123-4"><a href="#cb123-4" tabindex="-1"></a>##  [<span class="dv">7</span>] <span class="st">"CD8+ T Cells"</span>                 <span class="st">"B Cells"</span>                     </span>
<span id="cb123-5"><a href="#cb123-5" tabindex="-1"></a>##  [<span class="dv">9</span>] <span class="st">"CD4+ T Cells"</span>                 <span class="st">"PBMC Monocytes"</span></span></code></pre></div>
</div>
<div class="tab-pane" id="anndata-10" role="tabpanel" aria-labelledby="anndata-10-tab">

<div class="sourceCode" id="cb124"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/unique_values.html">unique_values</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="../reference/AML_h5ad.html">AML_h5ad</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  var <span class="op">=</span> <span class="st">"condensed_cell_type"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre class="anndata"><code><span><span class="co">##  [1] Plasma cells                 Primitive                   </span></span>
<span><span class="co">##  [3] Dendritic cells              Plasmacytoid dendritic cells</span></span>
<span><span class="co">##  [5] BM Monocytes                 NK Cells                    </span></span>
<span><span class="co">##  [7] CD8+ T Cells                 B Cells                     </span></span>
<span><span class="co">##  [9] CD4+ T Cells                 PBMC Monocytes              </span></span>
<span><span class="co">## 10 Levels: B Cells BM Monocytes CD4+ T Cells CD8+ T Cells ... Primitive</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="data-visualization">Data Visualization<a class="anchor" aria-label="anchor" href="#data-visualization"></a>
</h2>
<p>The data access functions mentioned in the previous section can be
applied downstream to create new plotting functions for single-cell
data. Functions created with SCUBA generics will be flexible across
object classes by default.</p>
<p>The general workflow of single-cell data visualization functions is
as follows:</p>
<ul>
<li>Data access</li>
<li>If applicable, statistics (average expression, percent expression,
number of cells in group, etc.)</li>
<li>Visualization code</li>
</ul>
<p>This section will demonstrate how to use SCUBA functions to simplify
data access. In the example below, a function will be created to plot
density plots for a feature. The function will accept a feature and a
metadata variable and will fetch data for the entered feature or
variable via SCUBA access functions. Plotting code downstream will
create the density plots.</p>
<p>The example below will follow principles of incremental development.
Each code block will show a “snapshot” of the development process, where
a progressively complex output will be created. Some intermediates are
ugly and/or difficult to interpret, but they will be corrected in
subsequent steps.</p>
<p>First off, we define the parameters to use for the function. To
facilitate development, the parameters are initialized at the top of the
script as variables. The function will be initially written as a script
that uses these variables, and the code beneath these variable
declarations will be moved into a function at the end of the
process.</p>
<p>The variables initialized include the object to use, a feature to
enter for the plot, and a metadata variable. These are set to the test
object provided with the SCUBA package, and a feature and metadata
variable present in that object. It does not matter which value is used
here as long as it exists. When developing your own functions, these
values could reflect a test case for the function you are
developing.</p>
<div class="sourceCode" id="cb126"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">object</span> <span class="op">&lt;-</span> <span class="va">AML_Seurat</span></span>
<span><span class="va">feature</span> <span class="op">&lt;-</span> <span class="st">"rna_GAPDH"</span></span>
<span><span class="va">metadata</span> <span class="op">&lt;-</span> <span class="st">"condensed_cell_type"</span></span></code></pre></div>
<div class="sourceCode" id="cb127"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">object</span> <span class="op">&lt;-</span> <span class="va">AML_Seurat</span></span>
<span><span class="va">feature</span> <span class="op">&lt;-</span> <span class="st">"rna_GAPDH"</span></span>
<span><span class="va">metadata</span> <span class="op">&lt;-</span> <span class="st">"condensed_cell_type"</span></span>
<span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="../reference/fetch_data.html">fetch_data</a></span><span class="op">(</span></span>
<span>    <span class="va">object</span>,</span>
<span>    <span class="co"># variables to fetch, the feature and metadata variable</span></span>
<span>    vars <span class="op">=</span> </span>
<span>      <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">feature</span>,</span>
<span>        <span class="va">metadata</span></span>
<span>        <span class="op">)</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
<p>After fetching the data, we will make a basic density plot using
<code><a href="https://ggplot2.tidyverse.org/reference/geom_density.html" class="external-link">geom_density()</a></code> from the ggplot2 package. In this stage, it
will be for the full object and not split by a metadata variable.</p>
<p>Make sure you have loaded ggplot2 before proceeding.</p>
<div class="sourceCode" id="cb128"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://ggplot2.tidyverse.org" class="external-link">"ggplot2"</a></span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"ggplot2"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span></code></pre></div>
<p>The use of <code>.data[[]]</code> in the <code><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes()</a></code>
specification allows us to supply the feature as a character vector to
ggplot. Using <code>.data[[]]</code> allows the script to accept any
feature without needing to make changes to the code, or having to rename
the column corresponding to the feature in the output of
<code><a href="../reference/fetch_data.html">fetch_data()</a></code>.</p>
<div class="sourceCode" id="cb129"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">object</span> <span class="op">&lt;-</span> <span class="va">AML_Seurat</span></span>
<span><span class="va">feature</span> <span class="op">&lt;-</span> <span class="st">"rna_GAPDH"</span></span>
<span><span class="va">metadata</span> <span class="op">&lt;-</span> <span class="st">"condensed_cell_type"</span></span>
<span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="../reference/fetch_data.html">fetch_data</a></span><span class="op">(</span></span>
<span>    <span class="va">object</span>,</span>
<span>    <span class="co"># variables to fetch, the feature and metadata variable</span></span>
<span>    vars <span class="op">=</span> </span>
<span>      <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">feature</span>,</span>
<span>        <span class="va">metadata</span></span>
<span>        <span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span></span>
<span><span class="va">plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span></span>
<span>    data <span class="op">=</span> <span class="va">data</span>,</span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">.data</span><span class="op">[[</span><span class="va">feature</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html" class="external-link">geom_density</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">plot</span></span></code></pre></div>
<p><img src="user_guide_files/figure-html/unnamed-chunk-51-1.png" width="700"></p>
<p>Next, we will add a fill attribute based on the metadata variable
entered. The metadata variable is supplied to the fill attribute, again
using <code>.data[[]]</code>.</p>
<div class="sourceCode" id="cb130"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">object</span> <span class="op">&lt;-</span> <span class="va">AML_Seurat</span></span>
<span><span class="va">feature</span> <span class="op">&lt;-</span> <span class="st">"rna_GAPDH"</span></span>
<span><span class="va">metadata</span> <span class="op">&lt;-</span> <span class="st">"condensed_cell_type"</span></span>
<span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="../reference/fetch_data.html">fetch_data</a></span><span class="op">(</span></span>
<span>    <span class="va">object</span>,</span>
<span>    <span class="co"># variables to fetch, the feature and metadata variable</span></span>
<span>    vars <span class="op">=</span> </span>
<span>      <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">feature</span>,</span>
<span>        <span class="va">metadata</span></span>
<span>        <span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span></span>
<span><span class="va">plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span></span>
<span>    data <span class="op">=</span> <span class="va">data</span>,</span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">.data</span><span class="op">[[</span><span class="va">feature</span><span class="op">]</span><span class="op">]</span>, fill <span class="op">=</span> <span class="va">.data</span><span class="op">[[</span><span class="va">metadata</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html" class="external-link">geom_density</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">plot</span></span></code></pre></div>
<p><img src="user_guide_files/figure-html/unnamed-chunk-52-1.png" width="700"></p>
<p>For this plot to be interperetable, the curves will need to be
“split” out into separate plots. This is done with
<code><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap()</a></code>.</p>
<div class="sourceCode" id="cb131"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">object</span> <span class="op">&lt;-</span> <span class="va">AML_Seurat</span></span>
<span><span class="va">feature</span> <span class="op">&lt;-</span> <span class="st">"rna_GAPDH"</span></span>
<span><span class="va">metadata</span> <span class="op">&lt;-</span> <span class="st">"condensed_cell_type"</span></span>
<span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="../reference/fetch_data.html">fetch_data</a></span><span class="op">(</span></span>
<span>    <span class="va">object</span>,</span>
<span>    <span class="co"># variables to fetch, the feature and metadata variable</span></span>
<span>    vars <span class="op">=</span> </span>
<span>      <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">feature</span>,</span>
<span>        <span class="va">metadata</span></span>
<span>        <span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span></span>
<span><span class="va">plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span></span>
<span>    data <span class="op">=</span> <span class="va">data</span>,</span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">.data</span><span class="op">[[</span><span class="va">feature</span><span class="op">]</span><span class="op">]</span>, fill <span class="op">=</span> <span class="va">.data</span><span class="op">[[</span><span class="va">metadata</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html" class="external-link">geom_density</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/vars.html" class="external-link">vars</a></span><span class="op">(</span><span class="va">.data</span><span class="op">[[</span><span class="va">metadata</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>,</span>
<span>        <span class="co"># Set a two-column layout</span></span>
<span>        ncol <span class="op">=</span> <span class="fl">2</span></span>
<span>    <span class="op">)</span></span>
<span></span>
<span><span class="va">plot</span></span></code></pre></div>
<p><img src="user_guide_files/figure-html/unnamed-chunk-53-1.png" width="700"></p>
<p>Next, plot visuals are improved. <code><a href="https://wilkelab.org/cowplot/reference/theme_cowplot.html" class="external-link">theme_cowplot()</a></code> from
the <code>cowplot</code> package is applied. This removes grids from the
plot region, and adds axes on the left and bottom of the plot.</p>
<p>The following additional changes will be added with a custom
theme:</p>
<ul>
<li>Labels for each facet are horizontally centered.</li>
<li>The legend is removed from the plot.</li>
<li>The gray box behind facet labels is made transparent using
<code>strip.background = element_rect(fill = "#FFFFFF00")</code>. The
last two digits in the fill color “#FFFFFF00” make the background
transparent by setting the <a href="https://www.digitalocean.com/community/tutorials/css-hex-code-colors-alpha-values#adding-an-alpha-value-to-css-hex-codes" class="external-link">alpha
value of the hex code color</a> to zero.</li>
<li>The text is made bold and shrank via <code>strip.text</code>.</li>
</ul>
<div class="sourceCode" id="cb132"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://wilkelab.org/cowplot/" class="external-link">"cowplot"</a></span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"cowplot"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://wilkelab.org/cowplot/" class="external-link">cowplot</a></span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb133"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">object</span> <span class="op">&lt;-</span> <span class="va">AML_Seurat</span></span>
<span><span class="va">feature</span> <span class="op">&lt;-</span> <span class="st">"rna_GAPDH"</span></span>
<span><span class="va">metadata</span> <span class="op">&lt;-</span> <span class="st">"condensed_cell_type"</span></span>
<span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="../reference/fetch_data.html">fetch_data</a></span><span class="op">(</span></span>
<span>    <span class="va">object</span>,</span>
<span>    <span class="co"># variables to fetch, the feature and metadata variable</span></span>
<span>    vars <span class="op">=</span> </span>
<span>      <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">feature</span>,</span>
<span>        <span class="va">metadata</span></span>
<span>        <span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span></span>
<span><span class="va">plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span></span>
<span>    data <span class="op">=</span> <span class="va">data</span>,</span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">.data</span><span class="op">[[</span><span class="va">feature</span><span class="op">]</span><span class="op">]</span>, fill <span class="op">=</span> <span class="va">.data</span><span class="op">[[</span><span class="va">metadata</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html" class="external-link">geom_density</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span></span>
<span>      <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/vars.html" class="external-link">vars</a></span><span class="op">(</span><span class="va">.data</span><span class="op">[[</span><span class="va">metadata</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>,</span>
<span>      ncol <span class="op">=</span> <span class="fl">2</span></span>
<span>      <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">cowplot</span><span class="fu">::</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/theme_cowplot.html" class="external-link">theme_cowplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span></span>
<span>      plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span>,</span>
<span>      legend.position <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>      strip.background <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_rect</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"#FFFFFF00"</span><span class="op">)</span>,</span>
<span>      strip.text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>face <span class="op">=</span> <span class="st">"bold"</span>, size <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">rel</a></span><span class="op">(</span><span class="fl">0.8</span><span class="op">)</span><span class="op">)</span></span>
<span>      <span class="op">)</span></span>
<span></span>
<span><span class="va">plot</span></span></code></pre></div>
<p><img src="user_guide_files/figure-html/unnamed-chunk-55-1.png" width="700"></p>
<p>At this point, the plot is in a finished form and is ready to be
packaged in a function.</p>
<p>At this stage, simply transfer the variables defined at the beginning
to the function as parameters, and move the subsequent code into the
function.</p>
<div class="sourceCode" id="cb134"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Transfer variables to parameters</span></span>
<span><span class="co"># object &lt;- AML_Seurat</span></span>
<span><span class="co"># feature &lt;- "rna_GAPDH"</span></span>
<span><span class="co"># metadata &lt;- "condensed_cell_type"</span></span>
<span></span>
<span><span class="va">density_plot</span> <span class="op">&lt;-</span></span>
<span>  <span class="kw">function</span><span class="op">(</span></span>
<span>    <span class="va">object</span>,</span>
<span>    <span class="va">feature</span>,</span>
<span>    <span class="va">metadata</span></span>
<span>    <span class="op">)</span><span class="op">{</span></span>
<span>    <span class="va">data</span> <span class="op">&lt;-</span> </span>
<span>      <span class="fu"><a href="../reference/fetch_data.html">fetch_data</a></span><span class="op">(</span></span>
<span>        <span class="va">object</span>,</span>
<span>        <span class="co"># variables to fetch, the feature and metadata variable</span></span>
<span>        vars <span class="op">=</span> </span>
<span>          <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">feature</span>,</span>
<span>            <span class="va">metadata</span></span>
<span>            <span class="op">)</span></span>
<span>        <span class="op">)</span></span>
<span>    </span>
<span>    <span class="va">plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span></span>
<span>        data <span class="op">=</span> <span class="va">data</span>,</span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">.data</span><span class="op">[[</span><span class="va">feature</span><span class="op">]</span><span class="op">]</span>, fill <span class="op">=</span> <span class="va">.data</span><span class="op">[[</span><span class="va">metadata</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span>        <span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html" class="external-link">geom_density</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span></span>
<span>          <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/vars.html" class="external-link">vars</a></span><span class="op">(</span><span class="va">.data</span><span class="op">[[</span><span class="va">metadata</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>,</span>
<span>          ncol <span class="op">=</span> <span class="fl">2</span></span>
<span>          <span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu">cowplot</span><span class="fu">::</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/theme_cowplot.html" class="external-link">theme_cowplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span></span>
<span>          plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span>,</span>
<span>          legend.position <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>          strip.background <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_rect</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"#FFFFFF00"</span><span class="op">)</span>,</span>
<span>          strip.text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>face <span class="op">=</span> <span class="st">"bold"</span>, size <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">rel</a></span><span class="op">(</span><span class="fl">0.8</span><span class="op">)</span><span class="op">)</span></span>
<span>          <span class="op">)</span></span>
<span>    </span>
<span>    <span class="va">plot</span></span>
<span>    <span class="op">}</span></span></code></pre></div>
<p>Now we call the function on each supported object class. The plots
have almost identical visuals, except for differences in assay names
when entering features.</p>
<div class="section level4 tabset tabset-pills">
<h4 class="tabset tabset-pills" id="section-16"></h4>



<ul class="nav nav-pills" id="section-16" role="tablist">
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#seurat-11" id="seurat-11-tab" type="button" role="tab" aria-controls="seurat-11" aria-selected="true" class="active nav-link">Seurat</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#singlecellexperiment-11" id="singlecellexperiment-11-tab" type="button" role="tab" aria-controls="singlecellexperiment-11" aria-selected="false" class="nav-link">SingleCellExperiment</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#anndata-11" id="anndata-11-tab" type="button" role="tab" aria-controls="anndata-11" aria-selected="false" class="nav-link">anndata</button></li>
</ul>
<div class="tab-content">
<div class="active  tab-pane" id="seurat-11" role="tabpanel" aria-labelledby="seurat-11-tab">

<div class="sourceCode" id="cb135"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">density_plot</span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">AML_Seurat</span>,</span>
<span>  feature <span class="op">=</span> <span class="st">"rna_GAPDH"</span>,</span>
<span>  metadata <span class="op">=</span> <span class="st">"condensed_cell_type"</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="user_guide_files/figure-html/unnamed-chunk-57-1.png" width="700"></p>
</div>
<div class="tab-pane" id="singlecellexperiment-11" role="tabpanel" aria-labelledby="singlecellexperiment-11-tab">

<div class="sourceCode" id="cb136"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">density_plot</span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="fu"><a href="../reference/AML_SCE.html">AML_SCE</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  feature <span class="op">=</span> <span class="st">"RNA_GAPDH"</span>,</span>
<span>  metadata <span class="op">=</span> <span class="st">"condensed_cell_type"</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="user_guide_files/figure-html/unnamed-chunk-58-1.png" width="700"></p>
</div>
<div class="tab-pane" id="anndata-11" role="tabpanel" aria-labelledby="anndata-11-tab">

<div class="sourceCode" id="cb137"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">density_plot</span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="fu"><a href="../reference/AML_h5ad.html">AML_h5ad</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  feature <span class="op">=</span> <span class="st">"X_GAPDH"</span>,</span>
<span>  metadata <span class="op">=</span> <span class="st">"condensed_cell_type"</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="user_guide_files/figure-html/unnamed-chunk-59-1.png" width="700"></p>
</div>
</div>
</div>
<div class="section level4 unnumbered">
<h4 class="unnumbered" id="section-17"></h4>
<p>The process described here can be applied to the creation of any
other single-cell visualization that uses expression data, metadata, or
reduction coordinates. SCUBA facilitates data access in visualization
functions, allowing users to access data through a call to an access
function instead of doing object-specific data wrangling.</p>
</div>
<div class="section level3">
<h3 id="utility-methods-for-visualization-functions">Utility Methods for Visualization Functions<a class="anchor" aria-label="anchor" href="#utility-methods-for-visualization-functions"></a>
</h3>
<p>SCUBA provides several utility methods for plotting functions that
define default properties, allowing users to optionally specify
reductions and layers to user for plotting without requiring users to
enter these parameters.</p>
<p>Below is an example feature plot function that uses the following
utility functions:</p>
<ul>
<li>
<code><a href="../reference/default_reduction.html">default_reduction()</a></code>: Returns the key of a default
reduction to use.</li>
<li>
<code><a href="../reference/default_layer.html">default_layer()</a></code>: Returns the conventional name of the
normalized layer in the object.</li>
<li>
<code><a href="../reference/get_all_cells.html">get_all_cells()</a></code>: Returns the IDs of all cells in the
object.</li>
<li>
<code><a href="../reference/reduction_dimnames.html">reduction_dimnames()</a></code>: Forms the names of reduction
dimensions to pull, for passage to <code>vars</code> in
<code>fetch_data</code>.</li>
</ul>
<div class="sourceCode" id="cb138"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">feature_plot</span> <span class="op">&lt;-</span> </span>
<span>  <span class="kw">function</span><span class="op">(</span></span>
<span>    <span class="va">object</span>,</span>
<span>    <span class="va">feature</span>,</span>
<span>    <span class="va">dims</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span>,</span>
<span>    <span class="va">reduction</span> <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>    <span class="va">layer</span> <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>    <span class="va">cells</span> <span class="op">=</span> <span class="cn">NULL</span></span>
<span>  <span class="op">)</span><span class="op">{</span></span>
<span>  <span class="co"># Determine a default reduction if the user does not enter one</span></span>
<span>  <span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NULL.html" class="external-link">is.null</a></span><span class="op">(</span><span class="va">reduction</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="va">reduction</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/default_reduction.html">default_reduction</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span>    <span class="op">}</span> </span>
<span>  </span>
<span>  <span class="co"># Define a default layer if undefined</span></span>
<span>  <span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NULL.html" class="external-link">is.null</a></span><span class="op">(</span><span class="va">layer</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="va">layer</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/default_layer.html">default_layer</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span>    <span class="op">}</span></span>
<span>  </span>
<span>  <span class="co"># Cells: if undefined, set to all cells in object</span></span>
<span>  <span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NULL.html" class="external-link">is.null</a></span><span class="op">(</span><span class="va">cells</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="va">cells</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_all_cells.html">get_all_cells</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span>    <span class="op">}</span></span>
<span>  </span>
<span>  <span class="co"># Form syntax for pulling reduction coordinates using the reduction </span></span>
<span>  <span class="co"># name and the dimensions entered in the function</span></span>
<span>  <span class="va">reduction_vars</span> <span class="op">&lt;-</span></span>
<span>    <span class="fu"><a href="../reference/reduction_dimnames.html">reduction_dimnames</a></span><span class="op">(</span></span>
<span>      object <span class="op">=</span> <span class="va">object</span>,</span>
<span>      reduction <span class="op">=</span> <span class="va">reduction</span>,</span>
<span>      dims <span class="op">=</span> <span class="va">dims</span></span>
<span>      <span class="op">)</span></span>
<span>  </span>
<span>  <span class="va">data</span> <span class="op">&lt;-</span></span>
<span>    <span class="fu"><a href="../reference/fetch_data.html">fetch_data</a></span><span class="op">(</span></span>
<span>      object <span class="op">=</span> <span class="va">object</span>,</span>
<span>      vars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">reduction_vars</span>, <span class="va">feature</span><span class="op">)</span>,</span>
<span>      layer <span class="op">=</span> <span class="va">layer</span>,</span>
<span>      cells <span class="op">=</span> <span class="va">cells</span></span>
<span>      <span class="op">)</span></span>
<span>  </span>
<span>  <span class="va">plot</span> <span class="op">&lt;-</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span></span>
<span>      <span class="va">data</span>, </span>
<span>      <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span></span>
<span>        x <span class="op">=</span> <span class="va">.data</span><span class="op">[[</span><span class="va">reduction_vars</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">]</span>, </span>
<span>        y <span class="op">=</span> <span class="va">.data</span><span class="op">[[</span><span class="va">reduction_vars</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">]</span>,</span>
<span>        color <span class="op">=</span> <span class="va">.data</span><span class="op">[[</span><span class="va">feature</span><span class="op">]</span><span class="op">]</span></span>
<span>        <span class="op">)</span></span>
<span>      <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">cowplot</span><span class="fu">::</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/theme_cowplot.html" class="external-link">theme_cowplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_color_gradientn</a></span><span class="op">(</span></span>
<span>      <span class="co"># Color cells using the viridis palette</span></span>
<span>      colors <span class="op">=</span> <span class="fu">viridisLite</span><span class="fu">::</span><span class="fu"><a href="https://sjmgarnier.github.io/viridisLite/reference/viridis.html" class="external-link">plasma</a></span><span class="op">(</span><span class="fl">12</span>, direction <span class="op">=</span> <span class="op">-</span><span class="fl">1</span><span class="op">)</span></span>
<span>      <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span></span>
<span>      plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span>,</span>
<span>      strip.background <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_rect</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"#FFFFFF00"</span><span class="op">)</span>,</span>
<span>      strip.text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>face <span class="op">=</span> <span class="st">"bold"</span>, size <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">rel</a></span><span class="op">(</span><span class="fl">0.8</span><span class="op">)</span><span class="op">)</span></span>
<span>      <span class="op">)</span></span>
<span>  </span>
<span>  <span class="va">plot</span></span>
<span>  <span class="op">}</span></span></code></pre></div>
<div class="section level4 tabset tabset-pills">
<h4 class="tabset tabset-pills" id="section-18"></h4>



<ul class="nav nav-pills" id="section-18" role="tablist">
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#seurat-12" id="seurat-12-tab" type="button" role="tab" aria-controls="seurat-12" aria-selected="true" class="active nav-link">Seurat</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#singlecellexperiment-12" id="singlecellexperiment-12-tab" type="button" role="tab" aria-controls="singlecellexperiment-12" aria-selected="false" class="nav-link">SingleCellExperiment</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#anndata-12" id="anndata-12-tab" type="button" role="tab" aria-controls="anndata-12" aria-selected="false" class="nav-link">anndata</button></li>
</ul>
<div class="tab-content">
<div class="active  tab-pane" id="seurat-12" role="tabpanel" aria-labelledby="seurat-12-tab">

<p>Running function without specifying default reduction, layer: utility
functions will supply defaults.</p>
<div class="sourceCode" id="cb139"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">feature_plot</span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">AML_Seurat</span>,</span>
<span>  feature <span class="op">=</span> <span class="st">"rna_GAPDH"</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="user_guide_files/figure-html/unnamed-chunk-61-1.png" width="700"></p>
<!--
Running the function with an alternate reduction:


```{.r .seurat}
feature_plot(
  object = AML_Seurat,
  feature = "rna_GAPDH",
  reduction = "pca"
  )
```
-->
<p>Running the function with the raw counts layer:</p>
<div class="sourceCode" id="cb140"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">feature_plot</span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">AML_Seurat</span>,</span>
<span>  feature <span class="op">=</span> <span class="st">"rna_GAPDH"</span>,</span>
<span>  layer <span class="op">=</span> <span class="st">"counts"</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="user_guide_files/figure-html/unnamed-chunk-63-1.png" width="700"></p>
</div>
<div class="tab-pane" id="singlecellexperiment-12" role="tabpanel" aria-labelledby="singlecellexperiment-12-tab">

<p>Running function without specifying default reduction, layer: utility
functions will supply defaults.</p>
<div class="sourceCode" id="cb141"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">feature_plot</span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="fu"><a href="../reference/AML_SCE.html">AML_SCE</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  feature <span class="op">=</span> <span class="st">"RNA_GAPDH"</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="user_guide_files/figure-html/unnamed-chunk-64-1.png" width="700"></p>
<!--Running the function with an alternate reduction:


```{.r .sce}
feature_plot(
  object = AML_SCE(),
  feature = "RNA_GAPDH",
  reduction = "PCA"
  )
```

-->
<p>Running the function using the raw counts layer:</p>
<div class="sourceCode" id="cb142"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">feature_plot</span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="fu"><a href="../reference/AML_SCE.html">AML_SCE</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  feature <span class="op">=</span> <span class="st">"RNA_GAPDH"</span>,</span>
<span>  layer <span class="op">=</span> <span class="st">"counts"</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="user_guide_files/figure-html/unnamed-chunk-66-1.png" width="700"></p>
</div>
<div class="tab-pane" id="anndata-12" role="tabpanel" aria-labelledby="anndata-12-tab">

<p>Running function without specifying default reduction, layer: utility
functions will supply defaults.</p>
<div class="sourceCode" id="cb143"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">feature_plot</span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="fu"><a href="../reference/AML_h5ad.html">AML_h5ad</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  feature <span class="op">=</span> <span class="st">"X_GAPDH"</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="user_guide_files/figure-html/unnamed-chunk-67-1.png" width="700"></p>
<!--
Running the function with an alternate reduction:


```{.r .anndata}
feature_plot(
  object = AML_h5ad(),
  feature = "X_GAPDH",
  reduction = "X_pca"
  )
```
-->
<p>Running the function with the raw counts layer:</p>
<div class="sourceCode" id="cb144"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">feature_plot</span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="fu"><a href="../reference/AML_h5ad.html">AML_h5ad</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  feature <span class="op">=</span> <span class="st">"X_GAPDH"</span>,</span>
<span>  layer <span class="op">=</span> <span class="st">"counts"</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="user_guide_files/figure-html/unnamed-chunk-69-1.png" width="700"></p>
</div>
</div>
</div>
<div class="section level4 unnumbered">
<h4 class="unnumbered" id="section-19"></h4>
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by First Last.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
